<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>skbio.metadata.Interval &#8212; scikit-bio 0.5.5 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/style.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="skbio.metadata.Interval.__eq__" href="skbio.metadata.Interval.__eq__.html" />
    <link rel="prev" title="Metadata (skbio.metadata)" href="../metadata.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head><body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.5.5</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../io.html">File I/O (<code class="docutils literal notranslate"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<code class="docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Alignments (<code class="docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<code class="docutils literal notranslate"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<code class="docutils literal notranslate"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<code class="docutils literal notranslate"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<code class="docutils literal notranslate"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../metadata.html">Metadata (<code class="docutils literal notranslate"><span class="pre">skbio.metadata</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<code class="docutils literal notranslate"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.metadata.Interval</a><ul>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="skbio-metadata-interval">
<h1>skbio.metadata.Interval<a class="headerlink" href="#skbio-metadata-interval" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="skbio.metadata.Interval">
<em class="property">class </em><code class="descclassname">skbio.metadata.</code><code class="descname">Interval</code><span class="sig-paren">(</span><em>interval_metadata</em>, <em>bounds</em>, <em>fuzzy=None</em>, <em>metadata=None</em><span class="sig-paren">)</span><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.5.5/skbio/metadata/_interval.py#L17"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.metadata.Interval" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores the bounds and metadata of an interval feature.</p>
<p>This class stores an interval feature. An interval feature
is defined as a sub-region of a biological sequence or sequence
alignment that is a functional entity, e.g., a gene, a riboswitch,
an exon, etc. It can span a single contiguous region or multiple
non-contiguous regions (e.g. multiple exons in a transcript, or
multiple genes in an operon).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>interval_metadata</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#object" title="(in Python v3.8)"><em>object</em></a>) – A reference to the <code class="docutils literal notranslate"><span class="pre">IntervalMetadata</span></code> object that this
<code class="docutils literal notranslate"><span class="pre">Interval</span></code> object is associated to.</li>
<li><strong>bounds</strong> (<em>iterable of tuple of int</em>) – Tuples representing start and end coordinates. It is <em>zero-based</em>
numbering. It is always inclusive on start bound and exclusive on
end bound.</li>
<li><strong>fuzzy</strong> (<em>iterable of tuple of bool</em><em>, </em><em>optional</em>) – Tuples representing the fuzziness of each bound coordinates.
If this isn’t specified, then the fuzziness of all bound
coordinates are <code class="docutils literal notranslate"><span class="pre">False</span></code>. If any of the coordinate fuzziness
is <code class="docutils literal notranslate"><span class="pre">True</span></code>, it indicates that the exact bound point of a
interval feature is unknown. The bound may begin or end at
some points outside the specified coordinates. This
accommodates the location format <a class="footnote-reference" href="#id2" id="id1">[1]</a> of INSDC.</li>
<li><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/dev/library/stdtypes.html#dict" title="(in Python v3.8)"><em>dict</em></a><em>, </em><em>optional</em>) – Dictionary of attributes storing information of the feature
such as “strand”, “gene_name”, or “product”.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="skbio.metadata.IntervalMetadata.html#skbio.metadata.IntervalMetadata" title="skbio.metadata.IntervalMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">skbio.metadata.IntervalMetadata</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>While the construction of an <code class="docutils literal notranslate"><span class="pre">Interval</span></code> object automatically add
itself to its associated <code class="docutils literal notranslate"><span class="pre">IntervalMetadata</span></code> object,
<code class="docutils literal notranslate"><span class="pre">IntervalMetadata.add</span></code> is the typical/easier way to
create and add it to <code class="docutils literal notranslate"><span class="pre">IntervalMetadata</span></code>.</p>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id2" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[1]</a></td><td><a class="reference external" href="ftp://ftp.ebi.ac.uk/pub/databases/embl/doc/FT_current.html#3.4.3">ftp://ftp.ebi.ac.uk/pub/databases/embl/doc/FT_current.html#3.4.3</a></td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>Hypothetically, let’s say we have a gene called “genA” with 10 nt
as shown in the following diagram.  The second row represents the
two exons (indicated by “=”) on this gene:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">TGGATTCTGC</span>
<span class="o">-====--==-</span>
<span class="mi">0123456789</span>
</pre></div>
</div>
<p>We can create an <code class="docutils literal notranslate"><span class="pre">Interval</span></code> object to represent the exons of the gene:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.metadata</span> <span class="k">import</span> <span class="n">Interval</span><span class="p">,</span> <span class="n">IntervalMetadata</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval_metadata</span> <span class="o">=</span> <span class="n">IntervalMetadata</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<p>Remember the coordinates are inclusive in lower bound and exclusive on
upper bound:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gene</span> <span class="o">=</span> <span class="n">Interval</span><span class="p">(</span><span class="n">interval_metadata</span><span class="p">,</span>
<span class="gp">... </span>                <span class="n">bounds</span><span class="o">=</span><span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">9</span><span class="p">)],</span>
<span class="gp">... </span>                <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;genA&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gene</span>    <span class="c1"># doctest: +ELLIPSIS</span>
<span class="go">Interval(interval_metadata=..., bounds=[(1, 5), (7, 9)], fuzzy=[(False, False), (False, False)], metadata={&#39;name&#39;: &#39;genA&#39;})</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">bounds</span></code></td>
<td>The coordinates of the interval feature.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">dropped</span></code></td>
<td>Boolean value indicating if the <code class="docutils literal notranslate"><span class="pre">Interval</span></code> object is dropped.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">fuzzy</span></code></td>
<td>The openness of each coordinate.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata</span></code></td>
<td>The metadata of the interval feature.</td>
</tr>
</tbody>
</table>
<p class="rubric">Built-ins</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.Interval.__eq__.html#skbio.metadata.Interval.__eq__" title="skbio.metadata.Interval.__eq__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interval1</span> <span class="pre">==</span> <span class="pre">interval2</span></code></a></td>
<td>Test if this <code class="docutils literal notranslate"><span class="pre">Interval</span></code> object is equal to another.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.metadata.Interval.__init_subclass__.html#skbio.metadata.Interval.__init_subclass__" title="skbio.metadata.Interval.__init_subclass__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__init_subclass__</span></code></a></td>
<td>This method is called when a class is subclassed.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.Interval.__ne__.html#skbio.metadata.Interval.__ne__" title="skbio.metadata.Interval.__ne__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interval1</span> <span class="pre">!=</span> <span class="pre">interval2</span></code></a></td>
<td>Test if this <code class="docutils literal notranslate"><span class="pre">Interval</span></code> object is not equal to another.</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.Interval.drop.html#skbio.metadata.Interval.drop" title="skbio.metadata.Interval.drop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">drop</span></code></a>()</td>
<td>Drop this <code class="docutils literal notranslate"><span class="pre">Interval</span></code> object from the interval metadata it links to.</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>