<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Sequences (skbio.sequence) &mdash; scikit-bio 0.4.0 documentation</title>
    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/style.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootswatch-3.3.4/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.4.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.4.0 documentation" href="index.html" />
    <link rel="next" title="skbio.sequence.Sequence" href="generated/skbio.sequence.Sequence.html" />
    <link rel="prev" title="skbio.io.util.open_files" href="generated/generated/skbio.io.util.open_files.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.4.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="io.html">File I/O (<tt class="docutils literal"><span class="pre">skbio.io</span></tt>)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Sequences (<tt class="docutils literal"><span class="pre">skbio.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="alignment.html">Alignments and Sequence collections (<tt class="docutils literal"><span class="pre">skbio.alignment</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="draw.html">Visualizations (<tt class="docutils literal"><span class="pre">skbio.draw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="diversity.html">Diversity calculations (<tt class="docutils literal"><span class="pre">skbio.diversity</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="stats.html">Statistics (<tt class="docutils literal"><span class="pre">skbio.stats</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="util.html">Utility functionality (<tt class="docutils literal"><span class="pre">skbio.util</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Sequences (<tt class="docutils literal"><span class="pre">skbio.sequence</span></tt>)</a><ul>
<li><a class="reference internal" href="#classes">Classes</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <span class="target" id="module-skbio.sequence"></span><div class="section" id="sequences-skbio-sequence">
<h1>Sequences (<a class="reference internal" href="#module-skbio.sequence" title="skbio.sequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence</span></tt></a>)<a class="headerlink" href="#sequences-skbio-sequence" title="Permalink to this headline">¶</a></h1>
<p>This module provides classes for storing and working with biological sequences,
including generic sequences which have no restrictions on which characters can
be included, and sequences based on IUPAC-defined sets of allowed characters
(including degenerate characters), including <tt class="docutils literal"><span class="pre">DNA</span></tt>, <tt class="docutils literal"><span class="pre">RNA</span></tt> and <tt class="docutils literal"><span class="pre">Protein</span></tt>
sequences. Common operations are defined as methods, for example computing the
reverse complement of a DNA sequence, or searching for N-glycosylation motifs
in <tt class="docutils literal"><span class="pre">Protein</span></tt> sequences. Class attributes are available to obtain valid
character sets, complement maps for different sequence types, and for obtaining
degenerate character definitions. Additionally this module defines the
<tt class="docutils literal"><span class="pre">GeneticCode</span></tt> class, which represents an immutable object that translates DNA
or RNA sequences into protein sequences.</p>
<p>The primary information stored for each different type of sequence object is
the underlying sequence data itself. This is stored as an immutable Numpy
array. Additionally, each type of sequence may include optional metadata
and positional metadata. Note that metadata and positional metadata are
mutable.</p>
<div class="section" id="classes">
<h2>Classes<a class="headerlink" href="#classes" title="Permalink to this headline">¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><tt class="xref py py-obj docutils literal"><span class="pre">Sequence</span></tt></a>(sequence[,&nbsp;metadata,&nbsp;...])</td>
<td>Store biological sequence data and optional associated metadata.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><tt class="xref py py-obj docutils literal"><span class="pre">DNA</span></tt></a>(sequence[,&nbsp;metadata,&nbsp;...])</td>
<td>Store DNA sequence data and optional associated metadata.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><tt class="xref py py-obj docutils literal"><span class="pre">RNA</span></tt></a>(sequence[,&nbsp;metadata,&nbsp;...])</td>
<td>Store RNA sequence data and optional associated metadata.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><tt class="xref py py-obj docutils literal"><span class="pre">Protein</span></tt></a>(sequence[,&nbsp;metadata,&nbsp;...])</td>
<td>Store protein sequence data and optional associated metadata.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/skbio.sequence.GeneticCode.html#skbio.sequence.GeneticCode" title="skbio.sequence.GeneticCode"><tt class="xref py py-obj docutils literal"><span class="pre">GeneticCode</span></tt></a>(amino_acids,&nbsp;starts[,&nbsp;name])</td>
<td>Genetic code for translating codons to amino acids.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">RNA</span>
</pre></div>
</div>
<p>New sequences are created with optional metadata and positional metadata
fields. Metadata is stored as a Python dict, while positional metadata
becomes a Pandas DataFrame.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s">&#39;ACCGGGTA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s">&#39;ACCGGGTA&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;id&#39;</span><span class="p">:</span><span class="s">&quot;my-sequence&quot;</span><span class="p">,</span> <span class="s">&#39;description&#39;</span><span class="p">:</span><span class="s">&quot;GFP&quot;</span><span class="p">},</span>
<span class="gp">... </span>         <span class="n">positional_metadata</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;quality&#39;</span><span class="p">:[</span><span class="mi">22</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">22</span><span class="p">,</span> <span class="mi">18</span><span class="p">,</span> <span class="mi">23</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">25</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s">&#39;ACCGGTA&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;id&#39;</span><span class="p">:</span><span class="s">&quot;my-sequence&quot;</span><span class="p">})</span>
</pre></div>
</div>
<p>New sequences can also be created from existing sequences, for example as their
reverse complement or degapped (i.e., unaligned) version.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s">&#39;.ACC--GGG-TA...&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;id&#39;</span><span class="p">:</span><span class="s">&#39;my-sequence&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span> <span class="o">=</span> <span class="n">d1</span><span class="o">.</span><span class="n">degap</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span>
<span class="go">DNA</span>
<span class="go">-----------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;id&#39;: &#39;my-sequence&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 8</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 62.50%</span>
<span class="go">-----------------------------</span>
<span class="go">0 ACCGGGTA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d3</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d3</span>
<span class="go">DNA</span>
<span class="go">-----------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;id&#39;: &#39;my-sequence&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 8</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 62.50%</span>
<span class="go">-----------------------------</span>
<span class="go">0 TACCCGGT</span>
</pre></div>
</div>
<p>It&#8217;s also straightforward to compute distances between sequences (optionally
using user-defined distance metrics, the default is Hamming distance which
requires that the sequences being compared are the same length) for use in
sequence clustering, phylogenetic reconstruction, etc.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">r1</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s">&#39;GACCCGCUUU&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r2</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s">&#39;GCCCCCCUUU&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r1</span><span class="o">.</span><span class="n">distance</span><span class="p">(</span><span class="n">r2</span><span class="p">)</span>
<span class="go">0.2</span>
</pre></div>
</div>
<p>Similarly, you can calculate the percent (dis)similarity between a pair of
aligned sequences.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">r3</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s">&#39;ACCGUUAGUC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r4</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s">&#39;ACGGGU--UC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r3</span><span class="o">.</span><span class="n">match_frequency</span><span class="p">(</span><span class="n">r4</span><span class="p">,</span> <span class="n">relative</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="go">0.6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r3</span><span class="o">.</span><span class="n">mismatch_frequency</span><span class="p">(</span><span class="n">r4</span><span class="p">,</span> <span class="n">relative</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="go">0.4</span>
</pre></div>
</div>
<p>Sequences can be searched for known motif types. This returns the slices that
describe the matches.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">r5</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s">&#39;AGG-GGACUGAA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_motifs</span><span class="p">(</span><span class="s">&#39;purine-run&#39;</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">motif</span>
<span class="go">slice(0, 3, None)</span>
<span class="go">slice(4, 7, None)</span>
<span class="go">slice(9, 12, None)</span>
</pre></div>
</div>
<p>Those slices can be used to extract the relevant subsequences.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_motifs</span><span class="p">(</span><span class="s">&#39;purine-run&#39;</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">r5</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="go">RNA</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 66.67%</span>
<span class="go">-----------------------------</span>
<span class="go">0 AGG</span>

<span class="go">RNA</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 66.67%</span>
<span class="go">-----------------------------</span>
<span class="go">0 GGA</span>

<span class="go">RNA</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">-----------------------------</span>
<span class="go">0 GAA</span>
</pre></div>
</div>
<p>And gaps or other features can be ignored while searching, as these may disrupt
otherwise meaningful motifs.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_motifs</span><span class="p">(</span><span class="s">&#39;purine-run&#39;</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ignore</span><span class="o">=</span><span class="n">r5</span><span class="o">.</span><span class="n">gaps</span><span class="p">()):</span>
<span class="gp">... </span>    <span class="n">r5</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="go">RNA</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 7</span>
<span class="go">    has gaps: True</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 66.67%</span>
<span class="go">-----------------------------</span>
<span class="go">0 AGG-GGA</span>

<span class="go">RNA</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">-----------------------------</span>
<span class="go">0 GAA</span>
</pre></div>
</div>
<p>In the above example, removing gaps from the resulting motif matches is easily
achieved, as the sliced matches themselves are sequences of the same type as
the input.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_motifs</span><span class="p">(</span><span class="s">&#39;purine-run&#39;</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ignore</span><span class="o">=</span><span class="n">r5</span><span class="o">.</span><span class="n">gaps</span><span class="p">()):</span>
<span class="gp">... </span>    <span class="n">r5</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span><span class="o">.</span><span class="n">degap</span><span class="p">()</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="go">RNA</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 6</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 66.67%</span>
<span class="go">-----------------------------</span>
<span class="go">0 AGGGGA</span>

<span class="go">RNA</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">-----------------------------</span>
<span class="go">0 GAA</span>
</pre></div>
</div>
<p>Sequences can similarly be searched for arbitrary patterns using regular
expressions.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_with_regex</span><span class="p">(</span><span class="s">&#39;(G+AC[UT])&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">match</span>
<span class="go">slice(4, 9, None)</span>
</pre></div>
</div>
<p>DNA can be transcribed to RNA:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s">&#39;ATGTGTATTTGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna</span>
<span class="go">RNA</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 12</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    GC-content: 25.00%</span>
<span class="go">-----------------------------</span>
<span class="go">0 AUGUGUAUUU GA</span>
</pre></div>
</div>
<p>Both DNA and RNA can be translated into a protein sequence. For example, let&#8217;s
translate our DNA and RNA sequences using NCBI&#8217;s standard genetic code (table
ID 1, the default genetic code in scikit-bio):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_dna</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_dna</span>
<span class="go">Protein</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    has stops: True</span>
<span class="go">-----------------------------</span>
<span class="go">0 MCI*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_rna</span> <span class="o">=</span> <span class="n">rna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_rna</span>
<span class="go">Protein</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    has stops: True</span>
<span class="go">-----------------------------</span>
<span class="go">0 MCI*</span>
</pre></div>
</div>
<p>The two translations are equivalent:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_dna</span> <span class="o">==</span> <span class="n">protein_from_rna</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Class-level methods contain information about the molecule types.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">DNA</span><span class="o">.</span><span class="n">degenerate_map</span><span class="p">[</span><span class="s">&#39;B&#39;</span><span class="p">]</span>
<span class="go">set([&#39;C&#39;, &#39;T&#39;, &#39;G&#39;])</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">RNA</span><span class="o">.</span><span class="n">degenerate_map</span><span class="p">[</span><span class="s">&#39;B&#39;</span><span class="p">]</span>
<span class="go">set([&#39;C&#39;, &#39;U&#39;, &#39;G&#39;])</span>
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>