<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.parse.sequences.parse_fasta &mdash; scikit-bio 0.1.4-dev documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.1.0/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1.4-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.1.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.1.4-dev documentation" href="../index.html" />
    <link rel="up" title="Parse biological sequences (skbio.parse.sequences)" href="../parse.sequences.html" />
    <link rel="next" title="skbio.parse.sequences.parse_fastq" href="skbio.parse.sequences.parse_fastq.html" />
    <link rel="prev" title="skbio.parse.sequences.parse_clustal" href="skbio.parse.sequences.parse_clustal.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.4-dev</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../core.alignment.html">Sequence collections and alignments (<tt class="docutils literal"><span class="pre">skbio.core.alignment</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.sequence.html">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.core.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.distance.html">Dissimilarity and distance matrices (<tt class="docutils literal"><span class="pre">skbio.core.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.exception.html">Exceptions (<tt class="docutils literal"><span class="pre">skbio.core.exception</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.warning.html">Warnings (<tt class="docutils literal"><span class="pre">skbio.core.warning</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.genetic_code.html">Genetic Code (<tt class="docutils literal"><span class="pre">skbio.core.genetic_code</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.core.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.core.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw.distributions.html">Distribution visualizations (<tt class="docutils literal"><span class="pre">skbio.draw.distributions</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../format.sequences.html">Format biological sequences (<tt class="docutils literal"><span class="pre">skbio.format.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.diversity.alpha.html">Alpha diversity measures (<tt class="docutils literal"><span class="pre">skbio.math.diversity.alpha</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.diversity.beta.html">Beta diversity measures (<tt class="docutils literal"><span class="pre">skbio.math.diversity.beta</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.distance.html">Distance-based statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.misc.html">Miscellaneous statistics utilities (<tt class="docutils literal"><span class="pre">skbio.math.stats.misc</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.ordination.html">Ordination methods (<tt class="docutils literal"><span class="pre">skbio.math.stats.ordination</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.spatial.html">Spatial Statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.spatial</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.gradient.html">Gradient analyses (<tt class="docutils literal"><span class="pre">skbio.math.gradient</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.subsample.html">Subsampling (<tt class="docutils literal"><span class="pre">skbio.math.subsample</span></tt>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../parse.sequences.html">Parse biological sequences (<tt class="docutils literal"><span class="pre">skbio.parse.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.io.html">I/O utils (<tt class="docutils literal"><span class="pre">skbio.util.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.misc.html">Miscellaneous Utilities (<tt class="docutils literal"><span class="pre">skbio.util.misc</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.testing.html">Testing utilities (<tt class="docutils literal"><span class="pre">skbio.util.testing</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.parse.sequences.parse_fasta</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="skbio-parse-sequences-parse-fasta">
<h1>skbio.parse.sequences.parse_fasta<a class="headerlink" href="#skbio-parse-sequences-parse-fasta" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="skbio.parse.sequences.parse_fasta">
<tt class="descclassname">skbio.parse.sequences.</tt><tt class="descname">parse_fasta</tt><big>(</big><em>infile</em>, <em>strict=True</em>, <em>label_to_name=None</em>, <em>finder=&lt;function parser at 0x27aeaa0&gt;</em>, <em>label_characters='&gt;'</em>, <em>ignore_comment=False</em><big>)</big><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/master/skbio/parse/sequences/fasta.py#L30"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.parse.sequences.parse_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator of labels and sequences from a fasta file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>infile</strong> : open file object or str</p>
<blockquote>
<div><p>An open fasta file or a path to a fasta file.</p>
</div></blockquote>
<p><strong>strict</strong> : bool</p>
<blockquote>
<div><p>If <tt class="docutils literal"><span class="pre">True</span></tt> a <tt class="docutils literal"><span class="pre">RecordError</span></tt> will be raised if there is a fasta label
line with no associated sequence, or a sequence with no associated
label line (in other words, if there is a partial record). If
<tt class="docutils literal"><span class="pre">False</span></tt>, partial records will be skipped.</p>
</div></blockquote>
<p><strong>label_to_name</strong> : function</p>
<blockquote>
<div><p>A function to apply to the sequence label (i.e., text on the header
line) before yielding it. By default, the sequence label is returned
with no processing. This function must take a single string as input
and return a single string as output.</p>
</div></blockquote>
<p><strong>finder</strong> : function</p>
<blockquote>
<div><p>The function to apply to find records in the fasta file. In general
you should not have to change this.</p>
</div></blockquote>
<p><strong>label_characters</strong> : str</p>
<blockquote>
<div><p>String used to indicate the beginning of a new record. In general you
should not have to change this.</p>
</div></blockquote>
<p><strong>ignore_comment</strong> : bool</p>
<blockquote>
<div><p>If <cite>True</cite>, split the sequence label on spaces, and return the label
only as the first space separated field (i.e., the sequence
identifier). Note: if both <tt class="docutils literal"><span class="pre">ignore_comment</span></tt> and <tt class="docutils literal"><span class="pre">label_to_name</span></tt> are
passed, <tt class="docutils literal"><span class="pre">ignore_comment</span></tt> is ignored (both operate on the label, so
there is potential for things to get messy otherwise).</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">two-item tuple of str</p>
<blockquote>
<div><p>yields the label and sequence for each entry.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>RecordError</strong></p>
<blockquote class="last">
<div><p>If <tt class="docutils literal"><span class="pre">strict</span> <span class="pre">==</span> <span class="pre">True</span></tt>, raises a <tt class="docutils literal"><span class="pre">RecordError</span></tt> if there is a fasta
label line with no associated sequence, or a sequence with no
associated label line (in other words, if there is a partial record).</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>Assume we have a fasta-formatted file with the following contents:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;seq1 db-accession-149855
CGATGTCGATCGATCGATCGATCAG
&gt;seq2 db-accession-34989
CATCGATCGATCGATGCATGCATGCATG
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">StringIO</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fasta_f</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="s">&#39;&gt;seq1 db-accession-149855</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                   <span class="s">&#39;CGATGTCGATCGATCGATCGATCAG</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                   <span class="s">&#39;&gt;seq2 db-accession-34989</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                   <span class="s">&#39;CATCGATCGATCGATGCATGCATGCATG</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can parse this as follows:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.parse.sequences</span> <span class="kn">import</span> <span class="n">parse_fasta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">label</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">parse_fasta</span><span class="p">(</span><span class="n">fasta_f</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>
<span class="go">seq1 db-accession-149855 CGATGTCGATCGATCGATCGATCAG</span>
<span class="go">seq2 db-accession-34989 CATCGATCGATCGATGCATGCATGCATG</span>
</pre></div>
</div>
<p>The sequence label or header line in a fasta file is defined as containing
two separate pieces of information, delimited by a space. The first space-
separated entry is the sequence identifier, and everything following the
first space is considered additional information (e.g., comments about the
source of the sequence or the molecule that it encodes). Often we don&#8217;t
care about that information within our code. If you want to just return the
sequence identifier from that line, you can pass <tt class="docutils literal"><span class="pre">ignore_comment=True</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">StringIO</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fasta_f</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="s">&#39;&gt;seq1 db-accession-149855</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                   <span class="s">&#39;CGATGTCGATCGATCGATCGATCAG</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                   <span class="s">&#39;&gt;seq2 db-accession-34989</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                   <span class="s">&#39;CATCGATCGATCGATGCATGCATGCATG</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.parse.sequences</span> <span class="kn">import</span> <span class="n">parse_fasta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">label</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">parse_fasta</span><span class="p">(</span><span class="n">fasta_f</span><span class="p">,</span> <span class="n">ignore_comment</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span>
<span class="go">seq1 CGATGTCGATCGATCGATCGATCAG</span>
<span class="go">seq2 CATCGATCGATCGATGCATGCATGCATG</span>
</pre></div>
</div>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>