<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.metadata.IntervalMetadata &#8212; scikit-bio 0.5.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.3.6/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.5.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="top" title="scikit-bio 0.5.1 documentation" href="../index.html" />
    <link rel="up" title="Metadata (skbio.metadata)" href="../metadata.html" />
    <link rel="next" title="Utility functionality (skbio.util)" href="../util.html" />
    <link rel="prev" title="skbio.metadata.Interval" href="skbio.metadata.Interval.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body role="document">
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.5.1</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../io.html">File I/O (<code class="docutils literal"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<code class="docutils literal"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Alignments (<code class="docutils literal"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<code class="docutils literal"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<code class="docutils literal"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<code class="docutils literal"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<code class="docutils literal"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../metadata.html">Metadata (<code class="docutils literal"><span class="pre">skbio.metadata</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<code class="docutils literal"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.metadata.IntervalMetadata</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="skbio-metadata-intervalmetadata">
<h1>skbio.metadata.IntervalMetadata<a class="headerlink" href="#skbio-metadata-intervalmetadata" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="skbio.metadata.IntervalMetadata">
<em class="property">class </em><code class="descclassname">skbio.metadata.</code><code class="descname">IntervalMetadata</code><span class="sig-paren">(</span><em>upper_bound</em><span class="sig-paren">)</span><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.5.1/skbio/metadata/_interval.py#L385"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.metadata.IntervalMetadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores the interval features.</p>
<p><code class="docutils literal"><span class="pre">IntervalMetadata</span></code> object allows storage, modification, and
querying of interval features covering a region of a single coordinate
system. For instance, this can be used to store functional annotations
about genes across a genome. This object is also applied to the sequence
alignment.</p>
<p>This object is typically coupled with another object, such as a
<code class="docutils literal"><span class="pre">Sequence</span></code> object (or its child class), or a <code class="docutils literal"><span class="pre">TabularMSA</span></code> object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>upper_bound</strong> : int</p>
<blockquote class="last">
<div><p>Defines the exclusive upper bound of the interval features. No
coordinate can be greater than it.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="skbio.metadata.Interval.html#skbio.metadata.Interval" title="skbio.metadata.Interval"><code class="xref py py-obj docutils literal"><span class="pre">skbio.metadata.Interval</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>This class stores coordinates of all feature bounds into a interval
tree. It allows the speed up of query-by-bound. The building of
interval tree is deferred until necessary to save computation. It is
updated from all coordinates only when you need to fetch info from
the interval tree.</p>
<p>When you add a method into this class and if you method need to fetch
info from <code class="docutils literal"><span class="pre">IntervalMetadata._interval_tree</span></code>, you should decorate it with
<code class="docutils literal"><span class="pre">_rebuild_tree</span></code>. This decorator will check if the current interval tree
is stale and will update it if so. Additionally, if your method add,
delete, or changes the coordinates of any interval features, you should
set <code class="docutils literal"><span class="pre">self._is_stale_tree</span></code> to <code class="docutils literal"><span class="pre">True</span></code> at the end of your method to
indicate the interval tree becomes stale.</p>
<p class="rubric">Examples</p>
<p>Let&#8217;s say we have a sequence of length 10 and want to add annotation
to it. Create an <code class="docutils literal"><span class="pre">IntervalMetadata</span></code> object:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.metadata</span> <span class="k">import</span> <span class="n">Interval</span><span class="p">,</span> <span class="n">IntervalMetadata</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">im</span> <span class="o">=</span> <span class="n">IntervalMetadata</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<p>Let&#8217;s add annotations of 3 genes:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">im</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">bounds</span><span class="o">=</span><span class="p">[(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">)],</span>
<span class="gp">... </span>       <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;sagB&#39;</span><span class="p">})</span>  
<span class="go">Interval(interval_metadata=..., bounds=[(3, 9)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagB&#39;})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">im</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">bounds</span><span class="o">=</span><span class="p">[(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">7</span><span class="p">)],</span>
<span class="gp">... </span>       <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;sagC&#39;</span><span class="p">})</span>  
<span class="go">Interval(interval_metadata=..., bounds=[(3, 7)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagC&#39;})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">im</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">bounds</span><span class="o">=</span><span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">)],</span>
<span class="gp">... </span>       <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;sagA&#39;</span><span class="p">})</span>  
<span class="go">Interval(interval_metadata=..., bounds=[(1, 2), (4, 7)], fuzzy=[(False, False), (False, False)], metadata={&#39;gene&#39;: &#39;sagA&#39;})</span>
</pre></div>
</div>
<p>Show the object representation:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">im</span>    
<span class="go">3 interval features</span>
<span class="go">-------------------</span>
<span class="go">Interval(interval_metadata=..., bounds=[(3, 9)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagB&#39;})</span>
<span class="go">Interval(interval_metadata=..., bounds=[(3, 7)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagC&#39;})</span>
<span class="go">Interval(interval_metadata=..., bounds=[(1, 2), (4, 7)], fuzzy=[(False, False), (False, False)], metadata={&#39;gene&#39;: &#39;sagA&#39;})</span>
</pre></div>
</div>
<p>We can sort the genes by their bounds:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">im</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">im</span>    
<span class="go">3 interval features</span>
<span class="go">-------------------</span>
<span class="go">Interval(interval_metadata=..., bounds=[(1, 2), (4, 7)], fuzzy=[(False, False), (False, False)], metadata={&#39;gene&#39;: &#39;sagA&#39;})</span>
<span class="go">Interval(interval_metadata=..., bounds=[(3, 7)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagC&#39;})</span>
<span class="go">Interval(interval_metadata=..., bounds=[(3, 9)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagB&#39;})</span>
</pre></div>
</div>
<p>Query the genes by bound and/or metadata:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">intvls</span> <span class="o">=</span> <span class="n">im</span><span class="o">.</span><span class="n">query</span><span class="p">([(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)],</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;foo&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">intvls</span><span class="p">)</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intvls</span> <span class="o">=</span> <span class="n">im</span><span class="o">.</span><span class="n">query</span><span class="p">([(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">9</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">intvls</span><span class="p">)</span>  
<span class="go">[Interval(interval_metadata=..., bounds=[(3, 9)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagB&#39;})]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intvls</span> <span class="o">=</span> <span class="n">im</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;sagA&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intvls</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">intvls</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intvls</span>  
<span class="go">[Interval(interval_metadata=..., bounds=[(1, 2), (4, 7)], fuzzy=[(False, False), (False, False)], metadata={&#39;gene&#39;: &#39;sagA&#39;})]</span>
</pre></div>
</div>
<p>Drop the gene(s) we get from query:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">im</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">intvls</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">im</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">im</span>   
<span class="go">2 interval features</span>
<span class="go">-------------------</span>
<span class="go">Interval(interval_metadata=..., bounds=[(3, 7)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagC&#39;})</span>
<span class="go">Interval(interval_metadata=..., bounds=[(3, 9)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagB&#39;})</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.lower_bound.html#skbio.metadata.IntervalMetadata.lower_bound" title="skbio.metadata.IntervalMetadata.lower_bound"><code class="xref py py-obj docutils literal"><span class="pre">lower_bound</span></code></a></td>
<td>The inclusive lower bound of interval features.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.num_interval_features.html#skbio.metadata.IntervalMetadata.num_interval_features" title="skbio.metadata.IntervalMetadata.num_interval_features"><code class="xref py py-obj docutils literal"><span class="pre">num_interval_features</span></code></a></td>
<td>The total number of interval features.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.upper_bound.html#skbio.metadata.IntervalMetadata.upper_bound" title="skbio.metadata.IntervalMetadata.upper_bound"><code class="xref py py-obj docutils literal"><span class="pre">upper_bound</span></code></a></td>
<td>The exclusive upper bound of interval features.</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.__copy__.html#skbio.metadata.IntervalMetadata.__copy__" title="skbio.metadata.IntervalMetadata.__copy__"><code class="xref py py-obj docutils literal"><span class="pre">copy.copy(im)</span></code></a></td>
<td>Return a shallow copy.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.__deepcopy__.html#skbio.metadata.IntervalMetadata.__deepcopy__" title="skbio.metadata.IntervalMetadata.__deepcopy__"><code class="xref py py-obj docutils literal"><span class="pre">copy.deepcopy(im)</span></code></a></td>
<td>Return a deep copy.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.__eq__.html#skbio.metadata.IntervalMetadata.__eq__" title="skbio.metadata.IntervalMetadata.__eq__"><code class="xref py py-obj docutils literal"><span class="pre">im1</span> <span class="pre">==</span> <span class="pre">im2</span></code></a></td>
<td>Test if this object is equal to another.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.__ne__.html#skbio.metadata.IntervalMetadata.__ne__" title="skbio.metadata.IntervalMetadata.__ne__"><code class="xref py py-obj docutils literal"><span class="pre">im1</span> <span class="pre">!=</span> <span class="pre">im2</span></code></a></td>
<td>Test if this object is not equal to another.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.add.html#skbio.metadata.IntervalMetadata.add" title="skbio.metadata.IntervalMetadata.add"><code class="xref py py-obj docutils literal"><span class="pre">add</span></code></a>(bounds[,&nbsp;fuzzy,&nbsp;metadata])</td>
<td>Create and add an <code class="docutils literal"><span class="pre">Interval</span></code> to this <code class="docutils literal"><span class="pre">IntervalMetadata</span></code>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.concat.html#skbio.metadata.IntervalMetadata.concat" title="skbio.metadata.IntervalMetadata.concat"><code class="xref py py-obj docutils literal"><span class="pre">concat</span></code></a>(interval_metadata)</td>
<td>Concatenate an iterable of <code class="docutils literal"><span class="pre">IntervalMetadata</span></code> objects.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.drop.html#skbio.metadata.IntervalMetadata.drop" title="skbio.metadata.IntervalMetadata.drop"><code class="xref py py-obj docutils literal"><span class="pre">drop</span></code></a>(intervals)</td>
<td>Drops Interval objects.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.merge.html#skbio.metadata.IntervalMetadata.merge" title="skbio.metadata.IntervalMetadata.merge"><code class="xref py py-obj docutils literal"><span class="pre">merge</span></code></a>(other)</td>
<td>Merge the interval features of another <code class="docutils literal"><span class="pre">IntervalMetadata</span></code> object.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.query.html#skbio.metadata.IntervalMetadata.query" title="skbio.metadata.IntervalMetadata.query"><code class="xref py py-obj docutils literal"><span class="pre">query</span></code></a>([bounds,&nbsp;metadata])</td>
<td>Yield <code class="docutils literal"><span class="pre">Interval</span></code> object with the bounds and attributes.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.metadata.IntervalMetadata.sort.html#skbio.metadata.IntervalMetadata.sort" title="skbio.metadata.IntervalMetadata.sort"><code class="xref py py-obj docutils literal"><span class="pre">sort</span></code></a>([ascending])</td>
<td>Sort interval features by their coordinates.</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5a0.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>