<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>skbio.alignment.global_pairwise_align_protein &#8212; scikit-bio 0.5.2 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.5.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="skbio.alignment.global_pairwise_align" href="skbio.alignment.global_pairwise_align.html" />
    <link rel="prev" title="skbio.alignment.global_pairwise_align_nucleotide" href="skbio.alignment.global_pairwise_align_nucleotide.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.5.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../io.html">File I/O (<code class="docutils literal"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<code class="docutils literal"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../alignment.html">Alignments (<code class="docutils literal"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<code class="docutils literal"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<code class="docutils literal"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<code class="docutils literal"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<code class="docutils literal"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../metadata.html">Metadata (<code class="docutils literal"><span class="pre">skbio.metadata</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<code class="docutils literal"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.alignment.global_pairwise_align_protein</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="skbio-alignment-global-pairwise-align-protein">
<h1>skbio.alignment.global_pairwise_align_protein<a class="headerlink" href="#skbio-alignment-global-pairwise-align-protein" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="skbio.alignment.global_pairwise_align_protein">
<code class="descclassname">skbio.alignment.</code><code class="descname">global_pairwise_align_protein</code><span class="sig-paren">(</span><em>seq1</em>, <em>seq2</em>, <em>gap_open_penalty=11</em>, <em>gap_extend_penalty=1</em>, <em>substitution_matrix=None</em>, <em>penalize_terminal_gaps=False</em><span class="sig-paren">)</span><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.5.2/skbio/alignment/_pairwise.py#L448"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.alignment.global_pairwise_align_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Globally align pair of protein seqs or alignments with Needleman-Wunsch</p>
<p>State: Experimental as of 0.4.0.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>seq1</strong> : Protein or TabularMSA[Protein]</p>
<blockquote>
<div><p>The first unaligned sequence(s).</p>
</div></blockquote>
<p><strong>seq2</strong> : Protein or TabularMSA[Protein]</p>
<blockquote>
<div><p>The second unaligned sequence(s).</p>
</div></blockquote>
<p><strong>gap_open_penalty</strong> : int or float, optional</p>
<blockquote>
<div><p>Penalty for opening a gap (this is substracted from previous best
alignment score, so is typically positive).</p>
</div></blockquote>
<p><strong>gap_extend_penalty</strong> : int or float, optional</p>
<blockquote>
<div><p>Penalty for extending a gap (this is substracted from previous best
alignment score, so is typically positive).</p>
</div></blockquote>
<p><strong>substitution_matrix: 2D dict (or similar), optional</strong></p>
<blockquote>
<div><p>Lookup for substitution scores (these values are added to the
previous best alignment score); default is BLOSUM 50.</p>
</div></blockquote>
<p><strong>penalize_terminal_gaps: bool, optional</strong></p>
<blockquote>
<div><p>If True, will continue to penalize gaps even after one sequence has
been aligned through its end. This behavior is true Needleman-Wunsch
alignment, but results in (biologically irrelevant) artifacts when
the sequences being aligned are of different length. This is <code class="docutils literal"><span class="pre">False</span></code>
by default, which is very likely to be the behavior you want in all or
nearly all cases.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">tuple</p>
<blockquote class="last">
<div><p><code class="docutils literal"><span class="pre">TabularMSA</span></code> object containing the aligned sequences, alignment score
(float), and start/end positions of each input sequence (iterable
of two-item tuples). Note that start/end positions are indexes into the
unaligned sequences.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="skbio.alignment.local_pairwise_align.html#skbio.alignment.local_pairwise_align" title="skbio.alignment.local_pairwise_align"><code class="xref py py-obj docutils literal"><span class="pre">local_pairwise_align</span></code></a>, <a class="reference internal" href="skbio.alignment.local_pairwise_align_protein.html#skbio.alignment.local_pairwise_align_protein" title="skbio.alignment.local_pairwise_align_protein"><code class="xref py py-obj docutils literal"><span class="pre">local_pairwise_align_protein</span></code></a>, <a class="reference internal" href="skbio.alignment.local_pairwise_align_nucleotide.html#skbio.alignment.local_pairwise_align_nucleotide" title="skbio.alignment.local_pairwise_align_nucleotide"><code class="xref py py-obj docutils literal"><span class="pre">local_pairwise_align_nucleotide</span></code></a>, <a class="reference internal" href="skbio.alignment.local_pairwise_align_ssw.html#skbio.alignment.local_pairwise_align_ssw" title="skbio.alignment.local_pairwise_align_ssw"><code class="xref py py-obj docutils literal"><span class="pre">skbio.alignment.local_pairwise_align_ssw</span></code></a>, <a class="reference internal" href="skbio.alignment.global_pairwise_align.html#skbio.alignment.global_pairwise_align" title="skbio.alignment.global_pairwise_align"><code class="xref py py-obj docutils literal"><span class="pre">global_pairwise_align</span></code></a>, <code class="xref py py-obj docutils literal"><span class="pre">global_pairwise_align_nucelotide</span></code></p>
</div>
<p class="rubric">Notes</p>
<p>Default <code class="docutils literal"><span class="pre">gap_open_penalty</span></code> and <code class="docutils literal"><span class="pre">gap_extend_penalty</span></code> parameters are
derived from the NCBI BLAST Server <a class="reference internal" href="#r205206" id="id1">[R205206]</a>.</p>
<p>The BLOSUM (blocks substitution matrices) amino acid substitution matrices
were originally defined in <a class="reference internal" href="#r206206" id="id2">[R206206]</a>.</p>
<p>This function can be use to align either a pair of sequences, a pair of
alignments, or a sequence and an alignment.</p>
<p class="rubric">References</p>
<table class="docutils citation" frame="void" id="r205206" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R205206]</td><td><em>(<a class="fn-backref" href="#id1">1</a>, <a class="fn-backref" href="#id3">2</a>)</em> <a class="reference external" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi">http://blast.ncbi.nlm.nih.gov/Blast.cgi</a></td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r206206" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R206206]</td><td><em>(<a class="fn-backref" href="#id2">1</a>, <a class="fn-backref" href="#id4">2</a>)</em> Amino acid substitution matrices from protein blocks.
S Henikoff and J G Henikoff.
Proc Natl Acad Sci U S A. Nov 15, 1992; 89(22): 10915-10919.</td></tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.5.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>