<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>skbio.sequence.GeneticCode &#8212; scikit-bio 0.5.2 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.5.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Sequence distance metrics (skbio.sequence.distance)" href="skbio.sequence.distance.html" />
    <link rel="prev" title="skbio.sequence.Protein" href="skbio.sequence.Protein.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.5.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../io.html">File I/O (<code class="docutils literal"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../sequence.html">Sequences (<code class="docutils literal"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Alignments (<code class="docutils literal"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<code class="docutils literal"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<code class="docutils literal"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<code class="docutils literal"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<code class="docutils literal"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../metadata.html">Metadata (<code class="docutils literal"><span class="pre">skbio.metadata</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<code class="docutils literal"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.sequence.GeneticCode</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="skbio-sequence-geneticcode">
<h1>skbio.sequence.GeneticCode<a class="headerlink" href="#skbio-sequence-geneticcode" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="skbio.sequence.GeneticCode">
<em class="property">class </em><code class="descclassname">skbio.sequence.</code><code class="descname">GeneticCode</code><span class="sig-paren">(</span><em>amino_acids</em>, <em>starts</em>, <em>name=''</em><span class="sig-paren">)</span><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.5.2/skbio/sequence/_genetic_code.py#L17"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.sequence.GeneticCode" title="Permalink to this definition">¶</a></dt>
<dd><p>Genetic code for translating codons to amino acids.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>amino_acids</strong> : consumable by <code class="docutils literal"><span class="pre">skbio.Protein</span></code> constructor</p>
<blockquote>
<div><p>64-character vector containing IUPAC amino acid characters. The order
of the amino acids should correspond to NCBI’s codon order (see <em>Notes</em>
section below). <cite>amino_acids</cite> is the “AAs” field in NCBI’s genetic
code format <a class="reference internal" href="#r356356" id="id1">[R356356]</a>.</p>
</div></blockquote>
<p><strong>starts</strong> : consumable by <code class="docutils literal"><span class="pre">skbio.Protein</span></code> constructor</p>
<blockquote>
<div><p>64-character vector containing only M and - characters, with start
codons indicated by M. The order of the amino acids should correspond
to NCBI’s codon order (see <em>Notes</em> section below). <cite>starts</cite> is the
“Starts” field in NCBI’s genetic code format <a class="reference internal" href="#r356356" id="id2">[R356356]</a>.</p>
</div></blockquote>
<p><strong>name</strong> : str, optional</p>
<blockquote class="last">
<div><p>Genetic code name. This is simply metadata and does not affect the
functionality of the genetic code itself.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="skbio.sequence.RNA.translate.html#skbio.sequence.RNA.translate" title="skbio.sequence.RNA.translate"><code class="xref py py-obj docutils literal"><span class="pre">RNA.translate</span></code></a>, <a class="reference internal" href="skbio.sequence.DNA.translate.html#skbio.sequence.DNA.translate" title="skbio.sequence.DNA.translate"><code class="xref py py-obj docutils literal"><span class="pre">DNA.translate</span></code></a>, <a class="reference internal" href="skbio.sequence.GeneticCode.from_ncbi.html#skbio.sequence.GeneticCode.from_ncbi" title="skbio.sequence.GeneticCode.from_ncbi"><code class="xref py py-obj docutils literal"><span class="pre">GeneticCode.from_ncbi</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>The genetic codes available via <code class="docutils literal"><span class="pre">GeneticCode.from_ncbi</span></code> and used
throughout the examples are defined in <a class="reference internal" href="#r356356" id="id3">[R356356]</a>. The genetic code strings
defined there are directly compatible with the <code class="docutils literal"><span class="pre">GeneticCode</span></code> constructor.</p>
<p>The order of <cite>amino_acids</cite> and <cite>starts</cite> should correspond to NCBI’s codon
order, defined in <a class="reference internal" href="#r356356" id="id4">[R356356]</a>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">UUUUUUUUUUUUUUUUCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG</span>
<span class="n">UUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGG</span>
<span class="n">UCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAG</span>
</pre></div>
</div>
<p>Note that scikit-bio displays this ordering using the IUPAC RNA alphabet,
while NCBI displays this same ordering using the IUPAC DNA alphabet (for
historical purposes).</p>
<p class="rubric">References</p>
<table class="docutils citation" frame="void" id="r356356" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R356356]</td><td><em>(<a class="fn-backref" href="#id1">1</a>, <a class="fn-backref" href="#id2">2</a>, <a class="fn-backref" href="#id3">3</a>, <a class="fn-backref" href="#id4">4</a>, <a class="fn-backref" href="#id5">5</a>)</em> <a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi</a></td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>Get NCBI’s standard genetic code (table ID 1, the default genetic code
in scikit-bio):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="k">import</span> <span class="n">GeneticCode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">GeneticCode</span><span class="o">.</span><span class="n">from_ncbi</span><span class="p">()</span>
<span class="go">GeneticCode (Standard)</span>
<span class="go">-------------------------------------------------------------------------</span>
<span class="go">  AAs  = FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG</span>
<span class="go">Starts = ---M---------------M---------------M----------------------------</span>
<span class="go">Base1  = UUUUUUUUUUUUUUUUCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG</span>
<span class="go">Base2  = UUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGG</span>
<span class="go">Base3  = UCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAG</span>
</pre></div>
</div>
<p>Get a different NCBI genetic code (25):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">GeneticCode</span><span class="o">.</span><span class="n">from_ncbi</span><span class="p">(</span><span class="mi">25</span><span class="p">)</span>
<span class="go">GeneticCode (Candidate Division SR1 and Gracilibacteria)</span>
<span class="go">-------------------------------------------------------------------------</span>
<span class="go">  AAs  = FFLLSSSSYY**CCGWLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG</span>
<span class="go">Starts = ---M-------------------------------M---------------M------------</span>
<span class="go">Base1  = UUUUUUUUUUUUUUUUCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG</span>
<span class="go">Base2  = UUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGG</span>
<span class="go">Base3  = UCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAG</span>
</pre></div>
</div>
<p>Define a custom genetic code:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">GeneticCode</span><span class="p">(</span><span class="s1">&#39;M&#39;</span> <span class="o">*</span> <span class="mi">64</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span> <span class="o">*</span> <span class="mi">64</span><span class="p">)</span>
<span class="go">GeneticCode</span>
<span class="go">-------------------------------------------------------------------------</span>
<span class="go">  AAs  = MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM</span>
<span class="go">Starts = ----------------------------------------------------------------</span>
<span class="go">Base1  = UUUUUUUUUUUUUUUUCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG</span>
<span class="go">Base2  = UUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGGUUUUCCCCAAAAGGGG</span>
<span class="go">Base3  = UCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAGUCAG</span>
</pre></div>
</div>
<p>Translate an RNA sequence to protein using NCBI’s standard genetic code:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="k">import</span> <span class="n">RNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;AUGCCACUUUAA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">GeneticCode</span><span class="o">.</span><span class="n">from_ncbi</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">rna</span><span class="p">)</span>
<span class="go">Protein</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    has stops: True</span>
<span class="go">--------------------------</span>
<span class="go">0 MPL*</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">name</span></code></td>
<td>Genetic code name.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal"><span class="pre">reading_frames</span></code></td>
<td>Six possible reading frames.</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.GeneticCode.__eq__.html#skbio.sequence.GeneticCode.__eq__" title="skbio.sequence.GeneticCode.__eq__"><code class="xref py py-obj docutils literal"><span class="pre">gc1</span> <span class="pre">==</span> <span class="pre">gc2</span></code></a></td>
<td>Determine if the genetic code is equal to another.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.GeneticCode.__ne__.html#skbio.sequence.GeneticCode.__ne__" title="skbio.sequence.GeneticCode.__ne__"><code class="xref py py-obj docutils literal"><span class="pre">gc1</span> <span class="pre">!=</span> <span class="pre">gc2</span></code></a></td>
<td>Determine if the genetic code is not equal to another.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.GeneticCode.__str__.html#skbio.sequence.GeneticCode.__str__" title="skbio.sequence.GeneticCode.__str__"><code class="xref py py-obj docutils literal"><span class="pre">str(gc)</span></code></a></td>
<td>Return string representation of the genetic code.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.GeneticCode.from_ncbi.html#skbio.sequence.GeneticCode.from_ncbi" title="skbio.sequence.GeneticCode.from_ncbi"><code class="xref py py-obj docutils literal"><span class="pre">from_ncbi</span></code></a>([table_id])</td>
<td>Return NCBI genetic code specified by table ID.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.GeneticCode.translate.html#skbio.sequence.GeneticCode.translate" title="skbio.sequence.GeneticCode.translate"><code class="xref py py-obj docutils literal"><span class="pre">translate</span></code></a>(sequence[,&nbsp;reading_frame,&nbsp;start,&nbsp;stop])</td>
<td>Translate RNA sequence into protein sequence.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.GeneticCode.translate_six_frames.html#skbio.sequence.GeneticCode.translate_six_frames" title="skbio.sequence.GeneticCode.translate_six_frames"><code class="xref py py-obj docutils literal"><span class="pre">translate_six_frames</span></code></a>(sequence[,&nbsp;start,&nbsp;stop])</td>
<td>Translate RNA into protein using six possible reading frames.</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.5.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>