<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>PHYLIP multiple sequence alignment format (skbio.io.format.phylip) &mdash; scikit-bio 0.4.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.3.4/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.4.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.4.2 documentation" href="../index.html" />
    <link rel="up" title="File I/O (skbio.io)" href="../io.html" />
    <link rel="next" title="QSeq format (skbio.io.format.qseq)" href="skbio.io.format.qseq.html" />
    <link rel="prev" title="Ordination results format (skbio.io.format.ordination)" href="skbio.io.format.ordination.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body role="document">
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.4.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../io.html">File I/O (<code class="docutils literal"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<code class="docutils literal"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Alignments (<code class="docutils literal"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<code class="docutils literal"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<code class="docutils literal"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<code class="docutils literal"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<code class="docutils literal"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<code class="docutils literal"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">PHYLIP multiple sequence alignment format (<code class="docutils literal"><span class="pre">skbio.io.format.phylip</span></code>)</a><ul>
<li><a class="reference internal" href="#format-support">Format Support</a></li>
<li><a class="reference internal" href="#format-specification">Format Specification</a><ul>
<li><a class="reference internal" href="#header-section">Header Section</a></li>
<li><a class="reference internal" href="#alignment-section">Alignment Section</a></li>
</ul>
</li>
<li><a class="reference internal" href="#format-parameters">Format Parameters</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <span class="target" id="module-skbio.io.format.phylip"></span><div class="section" id="phylip-multiple-sequence-alignment-format-skbio-io-format-phylip">
<h1>PHYLIP multiple sequence alignment format (<a class="reference internal" href="#module-skbio.io.format.phylip" title="skbio.io.format.phylip"><code class="xref py py-mod docutils literal"><span class="pre">skbio.io.format.phylip</span></code></a>)<a class="headerlink" href="#phylip-multiple-sequence-alignment-format-skbio-io-format-phylip" title="Permalink to this headline">¶</a></h1>
<p>The PHYLIP file format stores a multiple sequence alignment. The format was
originally defined and used in Joe Felsenstein&#8217;s PHYLIP package <a class="reference internal" href="#r172" id="id1">[R172]</a>, and has
since been supported by several other bioinformatics tools (e.g., RAxML <a class="reference internal" href="#r173" id="id2">[R173]</a>).
See <a class="reference internal" href="#r174" id="id3">[R174]</a> for the original format description, and <a class="reference internal" href="#r175" id="id4">[R175]</a> and <a class="reference internal" href="#r176" id="id5">[R176]</a> for additional
descriptions.</p>
<p>An example PHYLIP-formatted file taken from <a class="reference internal" href="#r174" id="id6">[R174]</a>:</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span>      <span class="mi">5</span>    <span class="mi">42</span>
<span class="n">Turkey</span>    <span class="n">AAGCTNGGGC</span> <span class="n">ATTTCAGGGT</span> <span class="n">GAGCCCGGGC</span> <span class="n">AATACAGGGT</span> <span class="n">AT</span>
<span class="n">Salmo</span> <span class="n">gairAAGCCTTGGC</span> <span class="n">AGTGCAGGGT</span> <span class="n">GAGCCGTGGC</span> <span class="n">CGGGCACGGT</span> <span class="n">AT</span>
<span class="n">H</span><span class="o">.</span> <span class="n">SapiensACCGGTTGGC</span> <span class="n">CGTTCAGGGT</span> <span class="n">ACAGGTTGGC</span> <span class="n">CGTTCAGGGT</span> <span class="n">AA</span>
<span class="n">Chimp</span>     <span class="n">AAACCCTTGC</span> <span class="n">CGTTACGCTT</span> <span class="n">AAACCGAGGC</span> <span class="n">CGGGACACTC</span> <span class="n">AT</span>
<span class="n">Gorilla</span>   <span class="n">AAACCCTTGC</span> <span class="n">CGGTACGCTT</span> <span class="n">AAACCATTGC</span> <span class="n">CGGTACGCTT</span> <span class="n">AA</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Original copyright notice for the above PHYLIP file:</p>
<p class="last"><em>(c) Copyright 1986-2008 by The University of Washington. Written by Joseph
Felsenstein. Permission is granted to copy this document provided that no
fee is charged for it and that this copyright notice is not removed.</em></p>
</div>
<div class="section" id="format-support">
<h2>Format Support<a class="headerlink" href="#format-support" title="Permalink to this headline">¶</a></h2>
<p><strong>Has Sniffer: Yes</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="8%" />
<col width="84%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Reader</th>
<th class="head">Writer</th>
<th class="head">Object Class</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Yes</td>
<td>Yes</td>
<td><a class="reference internal" href="skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-mod docutils literal"><span class="pre">skbio.alignment.TabularMSA</span></code></a></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="format-specification">
<h2>Format Specification<a class="headerlink" href="#format-specification" title="Permalink to this headline">¶</a></h2>
<p>PHYLIP format is a plain text format containing exactly two sections: a header
describing the dimensions of the alignment, followed by the multiple sequence
alignment itself.</p>
<p>The format described here is &#8220;strict&#8221; PHYLIP, as described in <a class="reference internal" href="#r175" id="id7">[R175]</a>. Strict
PHYLIP requires that each sequence identifier is exactly 10 characters long
(padded with spaces as necessary). Other bioinformatics tools (e.g., RAxML) may
relax this rule to allow for longer sequence identifiers. See the
<strong>Alignment Section</strong> below for more details.</p>
<p>The format described here is &#8220;sequential&#8221; format. The original PHYLIP format
specification <a class="reference internal" href="#r174" id="id8">[R174]</a> describes both sequential and interleaved formats.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">scikit-bio currently supports reading and writing strict, sequential
PHYLIP-formatted files. Relaxed and/or interleaved PHYLIP formats are not
supported.</p>
</div>
<div class="section" id="header-section">
<h3>Header Section<a class="headerlink" href="#header-section" title="Permalink to this headline">¶</a></h3>
<p>The header consists of a single line describing the dimensions of the
alignment. It <strong>must</strong> be the first line in the file. The header consists of
optional spaces, followed by two positive integers (<code class="docutils literal"><span class="pre">n</span></code> and <code class="docutils literal"><span class="pre">m</span></code>) separated
by one or more spaces. The first integer (<code class="docutils literal"><span class="pre">n</span></code>) specifies the number of
sequences (i.e., the number of rows) in the alignment. The second integer
(<code class="docutils literal"><span class="pre">m</span></code>) specifies the length of the sequences (i.e., the number of columns) in
the alignment. The smallest supported alignment dimensions are 1x1.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>scikit-bio will write the PHYLIP format header <em>without</em> preceding
spaces, and with only a single space between <code class="docutils literal"><span class="pre">n</span></code> and <code class="docutils literal"><span class="pre">m</span></code>.</p>
<p class="last">PHYLIP format <em>does not</em> support blank line(s) between the header and the
alignment.</p>
</div>
</div>
<div class="section" id="alignment-section">
<h3>Alignment Section<a class="headerlink" href="#alignment-section" title="Permalink to this headline">¶</a></h3>
<p>The alignment section immediately follows the header. It consists of <code class="docutils literal"><span class="pre">n</span></code>
lines (rows), one for each sequence in the alignment. Each row consists of a
sequence identifier (ID) and characters in the sequence, in fixed width format.</p>
<p>The sequence ID can be up to 10 characters long. IDs less than 10 characters
must have spaces appended to them to reach the 10 character fixed width. Within
an ID, all characters except newlines are supported, including spaces,
underscores, and numbers.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>When reading a PHYLIP-formatted file into an
<code class="docutils literal"><span class="pre">skbio.alignment.TabularMSA</span></code> object, sequence identifiers/labels are
stored as <code class="docutils literal"><span class="pre">TabularMSA</span></code> index labels (<code class="docutils literal"><span class="pre">index</span></code> property).</p>
<p>When writing an <code class="docutils literal"><span class="pre">skbio.alignment.TabularMSA</span></code> object as a PHYLIP-formatted
file, <code class="docutils literal"><span class="pre">TabularMSA</span></code> index labels will be converted to strings and written
as sequence identifiers/labels.</p>
<p class="last">scikit-bio supports the empty string (<code class="docutils literal"><span class="pre">''</span></code>) as a valid sequence ID. An
empty ID will be padded with 10 spaces when writing.</p>
</div>
<p>Sequence characters immediately follow the sequence ID. They <em>must</em> start at
the 11th character in the line, as the first 10 characters are reserved for the
sequence ID. While PHYLIP format does not explicitly restrict the set of
supported characters that may be used to represent a sequence, the original
format description <a class="reference internal" href="#r174" id="id9">[R174]</a> specifies the IUPAC nucleic acid lexicon for DNA or RNA
sequences, and the IUPAC protein lexicon for protein sequences. The original
PHYLIP specification uses <code class="docutils literal"><span class="pre">-</span></code> as a gap character, though older versions also
supported <code class="docutils literal"><span class="pre">.</span></code>. The sequence characters may contain optional spaces (e.g., to
improve readability), and both upper and lower case characters are supported.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>scikit-bio will read/write a PHYLIP-formatted file as long as the
alignment&#8217;s sequence characters are valid for the type of in-memory sequence
object being read into or written from. This differs from the PHYLIP
specification, which states that a PHYLIP-formatted file can only contain
valid IUPAC characters. See the <code class="docutils literal"><span class="pre">constructor</span></code> format parameter below for
details.</p>
<p>Since scikit-bio supports both <code class="docutils literal"><span class="pre">-</span></code> and <code class="docutils literal"><span class="pre">.</span></code> as gap characters (e.g., in
<code class="docutils literal"><span class="pre">DNA</span></code>, <code class="docutils literal"><span class="pre">RNA</span></code>, and <code class="docutils literal"><span class="pre">Protein</span></code> sequence objects), both are supported when
reading/writing a PHYLIP-formatted file.</p>
<p class="last">When writing a PHYLIP-formatted file, scikit-bio will split up each sequence
into chunks that are 10 characters long. Each chunk will be separated by a
single space. The sequence will always appear on a single line (sequential
format). It will <em>not</em> be wrapped across multiple lines. Sequences are
chunked in this manner for improved readability, and because most example
PHYLIP files are chunked in a similar way (e.g., see the example file
above). Note that this chunking is not required when reading
PHYLIP-formatted files, nor by the PHYLIP format specification itself.</p>
</div>
</div>
</div>
<div class="section" id="format-parameters">
<h2>Format Parameters<a class="headerlink" href="#format-parameters" title="Permalink to this headline">¶</a></h2>
<p>The only supported format parameter is <code class="docutils literal"><span class="pre">constructor</span></code>, which specifies the
type of in-memory sequence object to read each aligned sequence into. This must
be a subclass of <code class="docutils literal"><span class="pre">GrammaredSequence</span></code> (e.g., <code class="docutils literal"><span class="pre">DNA</span></code>, <code class="docutils literal"><span class="pre">RNA</span></code>, <code class="docutils literal"><span class="pre">Protein</span></code>)
and is a required format parameter. For example, if you know that the PHYLIP
file you&#8217;re reading contains DNA sequences, you would pass <code class="docutils literal"><span class="pre">constructor=DNA</span></code>
to the reader call.</p>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>Let&#8217;s create a <code class="docutils literal"><span class="pre">TabularMSA</span></code> with three DNA sequences:</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="k">import</span> <span class="n">TabularMSA</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACCGTTGTA-GTAGCT&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span><span class="s1">&#39;seq1&#39;</span><span class="p">}),</span>
<span class="gp">... </span>        <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;A--GTCGAA-GTACCT&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span><span class="s1">&#39;sequence-2&#39;</span><span class="p">}),</span>
<span class="gp">... </span>        <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AGAGTTGAAGGTATCT&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span><span class="s1">&#39;3&#39;</span><span class="p">})]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">minter</span><span class="o">=</span><span class="s1">&#39;id&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">----------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 3</span>
<span class="go">    position count: 16</span>
<span class="go">----------------------</span>
<span class="go">ACCGTTGTA-GTAGCT</span>
<span class="go">A--GTCGAA-GTACCT</span>
<span class="go">AGAGTTGAAGGTATCT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">Index([&#39;seq1&#39;, &#39;sequence-2&#39;, &#39;3&#39;], dtype=&#39;object&#39;)</span>
</pre></div>
</div>
<p>Now let&#8217;s write the <code class="docutils literal"><span class="pre">TabularMSA</span></code> to file in PHYLIP format and take a look at
the output:</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="k">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;phylip&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">3 16</span>
<span class="go">seq1      ACCGTTGTA- GTAGCT</span>
<span class="go">sequence-2A--GTCGAA- GTACCT</span>
<span class="go">3         AGAGTTGAAG GTATCT</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Notice that the 16-character sequences were split into two chunks, and that
each sequence appears on a single line (sequential format). Also note that each
sequence ID is padded with spaces to 10 characters in order to produce a fixed
width column.</p>
<p>If the index labels in a <code class="docutils literal"><span class="pre">TabularMSA</span></code> surpass the 10-character limit, an
error will be raised when writing:</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;long-sequence-2&#39;</span><span class="p">,</span> <span class="s1">&#39;seq3&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;phylip&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">skbio.io._exception.PhylipFormatError</span>: <span class="n">``TabularMSA`` can only be written in PHYLIP format if all sequence index labels have 10 or fewer characters. Found sequence with index label &#39;long-sequence-2&#39; that exceeds this limit. Use ``TabularMSA.reassign_index`` to assign shorter index labels.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>One way to work around this is to assign shorter index labels. The recommended
way to do this is via <code class="docutils literal"><span class="pre">TabularMSA.reassign_index</span></code>. For example, to reassign
default integer index labels:</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">reassign_index</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">Int64Index([0, 1, 2], dtype=&#39;int64&#39;)</span>
</pre></div>
</div>
<p>We can now write the <code class="docutils literal"><span class="pre">TabularMSA</span></code> in PHYLIP format:</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;phylip&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">3 16</span>
<span class="go">0         ACCGTTGTA- GTAGCT</span>
<span class="go">1         A--GTCGAA- GTACCT</span>
<span class="go">2         AGAGTTGAAG GTATCT</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<table class="docutils citation" frame="void" id="r172" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[R172]</a></td><td><a class="reference external" href="http://evolution.genetics.washington.edu/phylip.html">http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r173" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[R173]</a></td><td>RAxML Version 8: A tool for Phylogenetic Analysis and
Post-Analysis of Large Phylogenies&#8221;. In Bioinformatics, 2014</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r174" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R174]</td><td><em>(<a class="fn-backref" href="#id3">1</a>, <a class="fn-backref" href="#id6">2</a>, <a class="fn-backref" href="#id8">3</a>, <a class="fn-backref" href="#id9">4</a>)</em> <a class="reference external" href="http://evolution.genetics.washington.edu/phylip/doc/sequence.html">http://evolution.genetics.washington.edu/phylip/doc/sequence.html</a></td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r175" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R175]</td><td><em>(<a class="fn-backref" href="#id4">1</a>, <a class="fn-backref" href="#id7">2</a>)</em> <a class="reference external" href="http://www.phylo.org/tools/obsolete/phylip.html">http://www.phylo.org/tools/obsolete/phylip.html</a></td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r176" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id5">[R176]</a></td><td><a class="reference external" href="http://www.bioperl.org/wiki/PHYLIP_multiple_alignment_format">http://www.bioperl.org/wiki/PHYLIP_multiple_alignment_format</a></td></tr>
</tbody>
</table>
</div>
</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4a1+.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>