<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.core.genetic_code.GeneticCode &mdash; scikit-bio 0.1.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.1.0/spacelab/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.1.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.1.1 documentation" href="../index.html" />
    <link rel="up" title="Genetic Code (skbio.core.genetic_code)" href="../core.genetic_code.html" />
    <link rel="next" title="skbio.core.genetic_code.GeneticCode.blocks" href="skbio.core.genetic_code.GeneticCode.blocks.html" />
    <link rel="prev" title="skbio.core.genetic_code.GeneticCode.__eq__" href="skbio.core.genetic_code.GeneticCode.__eq__.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_white_ffffff.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.1</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../core.alignment.html">Sequence collections and alignments (<tt class="docutils literal"><span class="pre">skbio.core.alignment</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.sequence.html">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.core.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.distance.html">Dissimilarity and distance matrices (<tt class="docutils literal"><span class="pre">skbio.core.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.exception.html">Exceptions (<tt class="docutils literal"><span class="pre">skbio.core.exception</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.warning.html">Warnings (<tt class="docutils literal"><span class="pre">skbio.core.warning</span></tt>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../core.genetic_code.html">Genetic Code (<tt class="docutils literal"><span class="pre">skbio.core.genetic_code</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.core.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.core.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.ssw.html">Performing Striped Smith Waterman Alignments (<tt class="docutils literal"><span class="pre">skbio.core.ssw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw.distributions.html">Distribution visualizations (<tt class="docutils literal"><span class="pre">skbio.draw.distributions</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../format.sequences.html">Format biological sequences (<tt class="docutils literal"><span class="pre">skbio.format.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.diversity.alpha.html">Alpha diversity measures (<tt class="docutils literal"><span class="pre">skbio.math.diversity.alpha</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.distance.html">Distance-based statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.spatial.html">Spatial Statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.spatial</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.ordination.html">Ordination methods (<tt class="docutils literal"><span class="pre">skbio.math.stats.ordination</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.gradient.html">Gradient analyses (<tt class="docutils literal"><span class="pre">skbio.math.gradient</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.subsample.html">Subsampling (<tt class="docutils literal"><span class="pre">skbio.math.subsample</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../parse.sequences.html">Parse biological sequences (<tt class="docutils literal"><span class="pre">skbio.parse.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.io.html">I/O utils (<tt class="docutils literal"><span class="pre">skbio.util.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.misc.html">Miscellaneous Utilities (<tt class="docutils literal"><span class="pre">skbio.util.misc</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.testing.html">Testing utilities (<tt class="docutils literal"><span class="pre">skbio.util.testing</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.core.genetic_code.GeneticCode</a><ul>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="skbio-core-genetic-code-geneticcode">
<h1>skbio.core.genetic_code.GeneticCode<a class="headerlink" href="#skbio-core-genetic-code-geneticcode" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="skbio.core.genetic_code.GeneticCode">
<em class="property">class </em><tt class="descclassname">skbio.core.genetic_code.</tt><tt class="descname">GeneticCode</tt><big>(</big><em>code_sequence</em>, <em>id=None</em>, <em>name=None</em>, <em>start_codon_sequence=None</em><big>)</big><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.1.1/skbio/core/genetic_code.py#L94"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.core.genetic_code.GeneticCode" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to hold codon to amino acid mapping, and vice versa.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>code_sequence</strong> : str</p>
<blockquote>
<div><p>64-character string containing NCBI representation.</p>
</div></blockquote>
<p><strong>id</strong> : str, optional</p>
<blockquote>
<div><p>identifier for the object.</p>
</div></blockquote>
<p><strong>name</strong> : str, optional</p>
<blockquote>
<div><p>name for the object.</p>
</div></blockquote>
<p><strong>start_codon_sequence</strong> : str, optional</p>
<blockquote>
<div><p>starting point for the codon sequence.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">GeneticCode</p>
<blockquote>
<div><p>initialized <tt class="docutils literal"><span class="pre">GeneticCode</span></tt> object.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>GeneticCodeInitError</strong></p>
<blockquote class="last">
<div><p>If the length of <cite>code_sequence</cite> is different to <cite>64</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.core.genetic_code</span> <span class="kn">import</span> <span class="n">GeneticCode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sgc</span> <span class="o">=</span> <span class="n">GeneticCode</span><span class="p">(</span><span class="s">&#39;FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSR&#39;</span>
<span class="gp">... </span>                  <span class="s">&#39;RVVVVAAAADDEEGGGG&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><cite>*</cite> is used to denote termination as per the NCBI standard.
Although the genetic code objecs convert DNA to RNA and vice versa,
lists of codons that they produce will be provided in DNA format.</p>
</div>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.blocks.html#skbio.core.genetic_code.GeneticCode.blocks" title="skbio.core.genetic_code.GeneticCode.blocks"><tt class="xref py py-obj docutils literal"><span class="pre">blocks</span></tt></a></td>
<td>Returns list of lists of codon blocks in the genetic code.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="67%" />
<col width="33%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>code_sequence</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>id</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>name</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>start_codon_sequence</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>start_codons</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>codons</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>synonyms</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>sense_codons</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>anticodons</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.changes.html#skbio.core.genetic_code.GeneticCode.changes" title="skbio.core.genetic_code.GeneticCode.changes"><tt class="xref py py-obj docutils literal"><span class="pre">changes</span></tt></a>(other)</td>
<td>Returns dictionary of coddons that differ</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.get_stop_indices.html#skbio.core.genetic_code.GeneticCode.get_stop_indices" title="skbio.core.genetic_code.GeneticCode.get_stop_indices"><tt class="xref py py-obj docutils literal"><span class="pre">get_stop_indices</span></tt></a>(nucleotide_sequence[,&nbsp;start])</td>
<td>returns indexes for stop codons in the specified frame</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.is_start.html#skbio.core.genetic_code.GeneticCode.is_start" title="skbio.core.genetic_code.GeneticCode.is_start"><tt class="xref py py-obj docutils literal"><span class="pre">is_start</span></tt></a>(codon)</td>
<td>Checks if codon is a start codon</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.is_stop.html#skbio.core.genetic_code.GeneticCode.is_stop" title="skbio.core.genetic_code.GeneticCode.is_stop"><tt class="xref py py-obj docutils literal"><span class="pre">is_stop</span></tt></a>(codon)</td>
<td>Checks if codon is a stop codon</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.translate_six_frames.html#skbio.core.genetic_code.GeneticCode.translate_six_frames" title="skbio.core.genetic_code.GeneticCode.translate_six_frames"><tt class="xref py py-obj docutils literal"><span class="pre">translate_six_frames</span></tt></a>(nucleotide_sequence)</td>
<td>Translate nucleotide to protein sequences for all six reading frames</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.translate.html#skbio.core.genetic_code.GeneticCode.translate" title="skbio.core.genetic_code.GeneticCode.translate"><tt class="xref py py-obj docutils literal"><span class="pre">translate</span></tt></a>(nucleotide_sequence[,&nbsp;start])</td>
<td>Translate nucleotide to protein sequence</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.__repr__.html#skbio.core.genetic_code.GeneticCode.__repr__" title="skbio.core.genetic_code.GeneticCode.__repr__"><tt class="xref py py-obj docutils literal"><span class="pre">__repr__</span></tt></a>()</td>
<td>Returns reconstructable representation of the GeneticCode</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.__getitem__.html#skbio.core.genetic_code.GeneticCode.__getitem__" title="skbio.core.genetic_code.GeneticCode.__getitem__"><tt class="xref py py-obj docutils literal"><span class="pre">__getitem__</span></tt></a>(item)</td>
<td>Returns amino acid corresponding to codon, or codons for an aa.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.__str__.html#skbio.core.genetic_code.GeneticCode.__str__" title="skbio.core.genetic_code.GeneticCode.__str__"><tt class="xref py py-obj docutils literal"><span class="pre">__str__</span></tt></a>()</td>
<td>Returns code_sequence that constructs the GeneticCode</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.genetic_code.GeneticCode.__eq__.html#skbio.core.genetic_code.GeneticCode.__eq__" title="skbio.core.genetic_code.GeneticCode.__eq__"><tt class="xref py py-obj docutils literal"><span class="pre">__eq__</span></tt></a>(other)</td>
<td>Allows two GeneticCode objects to be compared to each other.</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>