<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.sequence.GeneticCode.translate &mdash; scikit-bio 0.4.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.3.4/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.4.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.4.0 documentation" href="../index.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.4.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../io.html">File I/O (<tt class="docutils literal"><span class="pre">skbio.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<tt class="docutils literal"><span class="pre">skbio.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Alignments and Sequence collections (<tt class="docutils literal"><span class="pre">skbio.alignment</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw.html">Visualizations (<tt class="docutils literal"><span class="pre">skbio.draw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<tt class="docutils literal"><span class="pre">skbio.diversity</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<tt class="docutils literal"><span class="pre">skbio.stats</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<tt class="docutils literal"><span class="pre">skbio.util</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.sequence.GeneticCode.translate</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="skbio-sequence-geneticcode-translate">
<h1>skbio.sequence.GeneticCode.translate<a class="headerlink" href="#skbio-sequence-geneticcode-translate" title="Permalink to this headline">¶</a></h1>
<dl class="method">
<dt id="skbio.sequence.GeneticCode.translate">
<tt class="descclassname">GeneticCode.</tt><tt class="descname">translate</tt><big>(</big><em>sequence</em>, <em>reading_frame=1</em>, <em>start='ignore'</em>, <em>stop='ignore'</em><big>)</big><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.4.0/skbio/sequence/_genetic_code.py#L403"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.sequence.GeneticCode.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate RNA sequence into protein sequence.</p>
<p>State: Stable as of 0.4.0.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>sequence</strong> : RNA</p>
<blockquote>
<div><p>RNA sequence to translate.</p>
</div></blockquote>
<p><strong>reading_frame</strong> : {1, 2, 3, -1, -2, -3}</p>
<blockquote>
<div><p>Reading frame to use in translation. 1, 2, and 3 are forward frames
and -1, -2, and -3 are reverse frames. If reverse (negative), will
reverse complement the sequence before translation.</p>
</div></blockquote>
<p><strong>start</strong> : {&#8216;ignore&#8217;, &#8216;require&#8217;, &#8216;optional&#8217;}</p>
<blockquote>
<div><p>How to handle start codons:</p>
<ul class="simple">
<li>&#8220;ignore&#8221;: translation will start from the beginning of the
reading frame, regardless of the presence of a start codon.</li>
<li>&#8220;require&#8221;: translation will start at the first start codon in
the reading frame, ignoring all prior positions. The first amino
acid in the translated sequence will <em>always</em> be methionine
(M character), even if an alternative start codon was used in
translation. This behavior most closely matches the underlying
biology since fMet doesn&#8217;t have a corresponding IUPAC character.
If a start codon does not exist, a <tt class="docutils literal"><span class="pre">ValueError</span></tt> is raised.</li>
<li>&#8220;optional&#8221;: if a start codon exists in the reading frame, matches
the behavior of &#8220;require&#8221;. If a start codon does not exist,
matches the behavior of &#8220;ignore&#8221;.</li>
</ul>
</div></blockquote>
<p><strong>stop</strong> : {&#8216;ignore&#8217;, &#8216;require&#8217;, &#8216;optional&#8217;}</p>
<blockquote>
<div><p>How to handle stop codons:</p>
<ul class="simple">
<li>&#8220;ignore&#8221;: translation will ignore the presence of stop codons and
translate to the end of the reading frame.</li>
<li>&#8220;require&#8221;: translation will terminate at the first stop codon.
The stop codon will not be included in the translated sequence.
If a stop codon does not exist, a <tt class="docutils literal"><span class="pre">ValueError</span></tt> is raised.</li>
<li>&#8220;optional&#8221;: if a stop codon exists in the reading frame, matches
the behavior of &#8220;require&#8221;. If a stop codon does not exist,
matches the behavior of &#8220;ignore&#8221;.</li>
</ul>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Protein</p>
<blockquote class="last">
<div><p>Translated sequence.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="skbio.sequence.GeneticCode.translate_six_frames.html#skbio.sequence.GeneticCode.translate_six_frames" title="skbio.sequence.GeneticCode.translate_six_frames"><tt class="xref py py-obj docutils literal"><span class="pre">translate_six_frames</span></tt></a></p>
</div>
<p class="rubric">Notes</p>
<p>Input RNA sequence metadata are included in the translated protein
sequence. Positional metadata are not included.</p>
<p class="rubric">Examples</p>
<p>Translate RNA into protein using NCBI&#8217;s standard genetic code (table ID
1, the default genetic code in scikit-bio):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">RNA</span><span class="p">,</span> <span class="n">GeneticCode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s">&#39;AGUAUUCUGCCACUGUAAGAA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sgc</span> <span class="o">=</span> <span class="n">GeneticCode</span><span class="o">.</span><span class="n">from_ncbi</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sgc</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">rna</span><span class="p">)</span>
<span class="go">Protein</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 7</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    has stops: True</span>
<span class="go">-----------------------------</span>
<span class="go">0 SILPL*E</span>
</pre></div>
</div>
<p>In this command, we used the default <tt class="docutils literal"><span class="pre">start</span></tt> behavior, which starts
translation at the beginning of the reading frame, regardless of the
presence of a start codon. If we specify &#8220;require&#8221;, translation will
start at the first start codon in the reading frame (in this example,
CUG), ignoring all prior positions:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sgc</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="s">&#39;require&#39;</span><span class="p">)</span>
<span class="go">Protein</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 5</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    has stops: True</span>
<span class="go">-----------------------------</span>
<span class="go">0 MPL*E</span>
</pre></div>
</div>
<p>Note that the codon coding for L (CUG) is an alternative start codon in
this genetic code. Since we specified &#8220;require&#8221; mode, methionine (M)
was used in place of the alternative start codon (L). This behavior
most closely matches the underlying biology since fMet doesn&#8217;t have a
corresponding IUPAC character.</p>
<p>Translate the same RNA sequence, also specifying that translation
terminate at the first stop codon in the reading frame:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sgc</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="s">&#39;require&#39;</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="s">&#39;require&#39;</span><span class="p">)</span>
<span class="go">Protein</span>
<span class="go">-----------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has non-degenerates: True</span>
<span class="go">    has stops: False</span>
<span class="go">-----------------------------</span>
<span class="go">0 MPL</span>
</pre></div>
</div>
<p>Passing &#8220;require&#8221; to both <tt class="docutils literal"><span class="pre">start</span></tt> and <tt class="docutils literal"><span class="pre">stop</span></tt> trims the translation
to the CDS (and in fact requires that one is present in the reading
frame). Changing the reading frame to 2 causes an exception to be
raised because a start codon doesn&#8217;t exist in the reading frame:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sgc</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="s">&#39;require&#39;</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="s">&#39;require&#39;</span><span class="p">,</span>
<span class="gp">... </span>              <span class="n">reading_frame</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> 
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">...</span>
</pre></div>
</div>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>