<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.parse.sequences.parse_fastq &mdash; scikit-bio 0.1.1-dev documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.1.0/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1.1-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.1.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.1.1-dev documentation" href="../index.html" />
    <link rel="up" title="Parse biological sequences (skbio.parse.sequences)" href="../parse.sequences.html" />
    <link rel="next" title="skbio.parse.sequences.parse_qual" href="skbio.parse.sequences.parse_qual.html" />
    <link rel="prev" title="skbio.parse.sequences.parse_fasta" href="skbio.parse.sequences.parse_fasta.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../test.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.1-dev</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../test.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../core.alignment.html">Sequence collections and alignments (<tt class="docutils literal"><span class="pre">skbio.core.alignment</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.sequence.html">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.core.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.distance.html">Dissimilarity and distance matrices (<tt class="docutils literal"><span class="pre">skbio.core.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.exception.html">Exceptions (<tt class="docutils literal"><span class="pre">skbio.core.exception</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.warning.html">Warnings (<tt class="docutils literal"><span class="pre">skbio.core.warning</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.genetic_code.html">Genetic Code (<tt class="docutils literal"><span class="pre">skbio.core.genetic_code</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.core.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.core.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.ssw.html">Performing Striped Smith Waterman Alignments (<tt class="docutils literal"><span class="pre">skbio.core.ssw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw.distributions.html">Distribution visualizations (<tt class="docutils literal"><span class="pre">skbio.draw.distributions</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../format.sequences.html">Format biological sequences (<tt class="docutils literal"><span class="pre">skbio.format.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.diversity.alpha.html">Alpha diversity measures (<tt class="docutils literal"><span class="pre">skbio.math.diversity.alpha</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.distance.html">Distance-based statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.spatial.html">Spatial Statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.spatial</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.ordination.html">Ordination methods (<tt class="docutils literal"><span class="pre">skbio.math.stats.ordination</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.gradient.html">Gradient analyses (<tt class="docutils literal"><span class="pre">skbio.math.gradient</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.subsample.html">Subsampling (<tt class="docutils literal"><span class="pre">skbio.math.subsample</span></tt>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../parse.sequences.html">Parse biological sequences (<tt class="docutils literal"><span class="pre">skbio.parse.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.io.html">I/O utils (<tt class="docutils literal"><span class="pre">skbio.util.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.misc.html">Miscellaneous Utilities (<tt class="docutils literal"><span class="pre">skbio.util.misc</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.testing.html">Testing utilities (<tt class="docutils literal"><span class="pre">skbio.util.testing</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.parse.sequences.parse_fastq</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="skbio-parse-sequences-parse-fastq">
<h1>skbio.parse.sequences.parse_fastq<a class="headerlink" href="#skbio-parse-sequences-parse-fastq" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="skbio.parse.sequences.parse_fastq">
<tt class="descclassname">skbio.parse.sequences.</tt><tt class="descname">parse_fastq</tt><big>(</big><em>data</em>, <em>strict=False</em>, <em>phred_offset=33</em><big>)</big><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/master/skbio/parse/sequences/fastq.py#L50"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.parse.sequences.parse_fastq" title="Permalink to this definition">¶</a></dt>
<dd><p>yields label, seq, and qual from a fastq file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>data</strong> : open file object or str</p>
<blockquote>
<div><p>An open fastq file (opened in binary mode) or a path to it.</p>
</div></blockquote>
<p><strong>strict</strong> : bool</p>
<blockquote>
<div><p>If strict is true a FastqParse error will be raised if the seq and qual
labels dont&#8217; match.</p>
</div></blockquote>
<p><strong>phred_offset</strong> : int or None</p>
<blockquote>
<div><p>Force a Phred offset, currently restricted to either 33 or 64.
Default behavior is to infer the Phred offset.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>label, seq, qual</strong> : (str, bytes, np.array)</p>
<blockquote class="last">
<div><p>yields the label, sequence and quality for each entry</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>Assume we have a fastq formatted file with the following contents:</p>
<div class="highlight-python"><div class="highlight"><pre>@seq1
AACACCAAACTTCTCCACCACGTGAGCTACAAAAG
+
````Y^T]`]c^cabcacc`^Lb^ccYT\T\Y\WF
@seq2
TATGTATATATAACATATACATATATACATACATA
+
]KZ[PY]_[YY^```ac^\\`bT``c`\aT``bbb
</pre></div>
</div>
<p>We can use the following code:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">StringIO</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.parse.sequences</span> <span class="kn">import</span> <span class="n">parse_fastq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fastq_f</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="s">&#39;@seq1</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                    <span class="s">&#39;AACACCAAACTTCTCCACCACGTGAGCTACAAAAG</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                    <span class="s">&#39;+</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                    <span class="s">&#39;````Y^T]`]c^cabcacc`^Lb^ccYT\T\Y\WF</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                    <span class="s">&#39;@seq2</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                    <span class="s">&#39;TATGTATATATAACATATACATATATACATACATA</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                    <span class="s">&#39;+</span><span class="se">\n</span><span class="s">&#39;</span>
<span class="gp">... </span>                    <span class="s">&#39;]KZ[PY]_[YY^```ac^</span><span class="se">\\</span><span class="s">\`bT``c`</span><span class="se">\\</span><span class="s">aT``bbb</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">label</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">qual</span> <span class="ow">in</span> <span class="n">parse_fastq</span><span class="p">(</span><span class="n">fastq_f</span><span class="p">,</span> <span class="n">phred_offset</span><span class="o">=</span><span class="mi">64</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">label</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">seq</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">qual</span>
<span class="go">seq1</span>
<span class="go">AACACCAAACTTCTCCACCACGTGAGCTACAAAAG</span>
<span class="go">[32 32 32 32 25 30 20 29 32 29 35 30 35 33 34 35 33 35 35 32 30 12 34 30 35</span>
<span class="go"> 35 25 20 28 20 28 25 28 23  6]</span>
<span class="go">seq2</span>
<span class="go">TATGTATATATAACATATACATATATACATACATA</span>
<span class="go">[29 11 26 27 16 25 29 31 27 25 25 30 32 32 32 33 35 30 28 28 32 34 20 32 32</span>
<span class="go"> 35 32 28 33 20 32 32 34 34 34]</span>
</pre></div>
</div>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>