<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.diversity.beta.weighted_unifrac &mdash; scikit-bio 0.4.2 documentation</title>
    
    <link rel="stylesheet" href="../../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/bootswatch-3.3.4/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.4.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../_static/copybutton.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../../_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.4.2 documentation" href="../../index.html" />
    <link rel="up" title="Beta diversity measures (skbio.diversity.beta)" href="../skbio.diversity.beta.html" />
    <link rel="next" title="skbio.diversity.alpha_diversity" href="../skbio.diversity.alpha_diversity.html" />
    <link rel="prev" title="skbio.diversity.beta.unweighted_unifrac" href="skbio.diversity.beta.unweighted_unifrac.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body role="document">
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.4.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../io.html">File I/O (<code class="docutils literal"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../sequence.html">Sequences (<code class="docutils literal"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../alignment.html">Alignments (<code class="docutils literal"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tree.html">Tree representations (<code class="docutils literal"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../workflow.html">Constructing workflows (<code class="docutils literal"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../diversity.html">Diversity calculations (<code class="docutils literal"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../stats.html">Statistics (<code class="docutils literal"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../util.html">Utility functionality (<code class="docutils literal"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.diversity.beta.weighted_unifrac</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="skbio-diversity-beta-weighted-unifrac">
<h1>skbio.diversity.beta.weighted_unifrac<a class="headerlink" href="#skbio-diversity-beta-weighted-unifrac" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="skbio.diversity.beta.weighted_unifrac">
<code class="descclassname">skbio.diversity.beta.</code><code class="descname">weighted_unifrac</code><span class="sig-paren">(</span><em>u_counts</em>, <em>v_counts</em>, <em>otu_ids</em>, <em>tree</em>, <em>normalized=False</em>, <em>validate=True</em><span class="sig-paren">)</span><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.4.2/skbio/diversity/beta/_unifrac.py#L154"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.diversity.beta.weighted_unifrac" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute weighted UniFrac with or without branch length normalization</p>
<p>State: Experimental as of 0.4.1.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>u_counts, v_counts: list, np.array</strong></p>
<blockquote>
<div><p>Vectors of counts/abundances of OTUs for two samples. Must be equal
length.</p>
</div></blockquote>
<p><strong>otu_ids: list, np.array</strong></p>
<blockquote>
<div><p>Vector of OTU ids corresponding to tip names in <code class="docutils literal"><span class="pre">tree</span></code>. Must be the
same length as <code class="docutils literal"><span class="pre">u_counts</span></code> and <code class="docutils literal"><span class="pre">v_counts</span></code>.</p>
</div></blockquote>
<p><strong>tree: skbio.TreeNode</strong></p>
<blockquote>
<div><p>Tree relating the OTUs in otu_ids. The set of tip names in the tree can
be a superset of <code class="docutils literal"><span class="pre">otu_ids</span></code>, but not a subset.</p>
</div></blockquote>
<p><strong>normalized: boolean, optional</strong></p>
<blockquote>
<div><p>If <code class="docutils literal"><span class="pre">True</span></code>, apply branch length normalization, which is described in
<a class="reference internal" href="#r56" id="id1">[R56]</a>. Resulting distances will then be in the range <code class="docutils literal"><span class="pre">[0,</span> <span class="pre">1]</span></code>.</p>
</div></blockquote>
<p><strong>validate: bool, optional</strong></p>
<blockquote>
<div><p>If <cite>False</cite>, validation of the input won&#8217;t be performed. This step can
be slow, so if validation is run elsewhere it can be disabled here.
However, invalid input data can lead to invalid results or error
messages that are hard to interpret, so this step should not be
bypassed if you&#8217;re not certain that your input data are valid. See
<a class="reference internal" href="../../diversity.html#module-skbio.diversity" title="skbio.diversity"><code class="xref py py-mod docutils literal"><span class="pre">skbio.diversity</span></code></a> for the description of what validation entails
so you can determine if you can safely disable validation.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">float</p>
<blockquote>
<div><p>The weighted UniFrac distance between the two samples.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>ValueError, MissingNodeError, DuplicateNodeError</strong></p>
<blockquote class="last">
<div><p>If validation fails. Exact error will depend on what was invalid.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="skbio.diversity.beta.unweighted_unifrac.html#skbio.diversity.beta.unweighted_unifrac" title="skbio.diversity.beta.unweighted_unifrac"><code class="xref py py-obj docutils literal"><span class="pre">unweighted_unifrac</span></code></a>, <a class="reference internal" href="../../diversity.html#module-skbio.diversity" title="skbio.diversity"><code class="xref py py-obj docutils literal"><span class="pre">skbio.diversity</span></code></a>, <a class="reference internal" href="../skbio.diversity.beta_diversity.html#skbio.diversity.beta_diversity" title="skbio.diversity.beta_diversity"><code class="xref py py-obj docutils literal"><span class="pre">skbio.diversity.beta_diversity</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>Weighted UniFrac was originally described in <a class="reference internal" href="#r56" id="id2">[R56]</a>, which includes a
discussion of unweighted (qualitative) versus weighted (quantitiative)
diversity metrics. Deeper mathemtical discussions of this metric is
presented in <a class="reference internal" href="#r57" id="id3">[R57]</a>.</p>
<p>If computing weighted UniFrac for multiple pairs of samples, using
<code class="docutils literal"><span class="pre">skbio.diversity.beta_diversity</span></code> will be much faster than calling this
function individually on each sample.</p>
<p>This implementation differs from that in PyCogent (and therefore QIIME
versions less than 2.0.0) by imposing a few additional restrictions on the
inputs. First, the input tree must be rooted. In PyCogent, if an unrooted
tree was provided that had a single trifurcating node (a newick convention
for unrooted trees) that node was considered the root of the tree. Next,
all OTU IDs must be tips in the tree. PyCogent would silently ignore OTU
IDs that were not present the tree. To reproduce UniFrac results from
PyCogent with scikit-bio, ensure that your PyCogent UniFrac calculations
are performed on a rooted tree and that all OTU IDs are present in the
tree.</p>
<p>This implementation of weighted UniFrac is the array-based implementation
described in <a class="reference internal" href="#r58" id="id4">[R58]</a>.</p>
<p class="rubric">References</p>
<table class="docutils citation" frame="void" id="r56" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R56]</td><td><em>(<a class="fn-backref" href="#id1">1</a>, <a class="fn-backref" href="#id2">2</a>, <a class="fn-backref" href="#id5">3</a>)</em> Lozupone, C. A., Hamady, M., Kelley, S. T. &amp; Knight, R. Quantitative
and qualitative beta diversity measures lead to different insights into
factors that structure microbial communities. Appl. Environ. Microbiol.
73, 1576-1585 (2007).</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r57" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R57]</td><td><em>(<a class="fn-backref" href="#id3">1</a>, <a class="fn-backref" href="#id6">2</a>)</em> Lozupone, C., Lladser, M. E., Knights, D., Stombaugh, J. &amp; Knight,
R. UniFrac: an effective distance metric for microbial community
comparison. ISME J. 5, 169-172 (2011).</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r58" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R58]</td><td><em>(<a class="fn-backref" href="#id4">1</a>, <a class="fn-backref" href="#id7">2</a>)</em> Hamady M, Lozupone C, Knight R. Fast UniFrac: facilitating high-
throughput phylogenetic analyses of microbial communities including
analysis of pyrosequencing and PhyloChip data.  ISME J. 4(1):17-27
(2010).</td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>Assume we have the following abundance data for two samples, <code class="docutils literal"><span class="pre">u</span></code> and
<code class="docutils literal"><span class="pre">v</span></code>, represented as a pair of counts vectors. These counts represent the
number of times specific Operational Taxonomic Units, or OTUs, were
observed in each of the samples.</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">u_counts</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">v_counts</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<p>Because UniFrac is a phylogenetic diversity metric, we need to know which
OTU each count corresponds to, which we&#8217;ll provide as <code class="docutils literal"><span class="pre">otu_ids</span></code>.</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">otu_ids</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;OTU1&#39;</span><span class="p">,</span> <span class="s1">&#39;OTU2&#39;</span><span class="p">,</span> <span class="s1">&#39;OTU3&#39;</span><span class="p">,</span> <span class="s1">&#39;OTU4&#39;</span><span class="p">,</span> <span class="s1">&#39;OTU5&#39;</span><span class="p">,</span> <span class="s1">&#39;OTU6&#39;</span><span class="p">,</span> <span class="s1">&#39;OTU7&#39;</span><span class="p">,</span>
<span class="gp">... </span>           <span class="s1">&#39;OTU8&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>We also need a phylogenetic tree that relates the OTUs to one another.</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="k">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="k">import</span> <span class="n">TreeNode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span>
<span class="gp">... </span>                     <span class="s1">u&#39;(((((OTU1:0.5,OTU2:0.5):0.5,OTU3:1.0):1.0):0.0,&#39;</span>
<span class="gp">... </span>                     <span class="s1">u&#39;(OTU4:0.75,(OTU5:0.5,((OTU6:0.33,OTU7:0.62):0.5&#39;</span>
<span class="gp">... </span>                     <span class="s1">u&#39;,OTU8:0.5):0.5):0.5):1.25):0.0)root;&#39;</span><span class="p">))</span>
</pre></div>
</div>
<p>Compute the weighted UniFrac distance between the samples.</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.diversity.beta</span> <span class="k">import</span> <span class="n">weighted_unifrac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">wu</span> <span class="o">=</span> <span class="n">weighted_unifrac</span><span class="p">(</span><span class="n">u_counts</span><span class="p">,</span> <span class="n">v_counts</span><span class="p">,</span> <span class="n">otu_ids</span><span class="p">,</span> <span class="n">tree</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">wu</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="go">1.54</span>
</pre></div>
</div>
<p>Compute the weighted UniFrac distance between the samples including
branch length normalization so the value falls in the range <code class="docutils literal"><span class="pre">[0.0,</span> <span class="pre">1.0]</span></code>.</p>
<div class="highlight-python3"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">wu</span> <span class="o">=</span> <span class="n">weighted_unifrac</span><span class="p">(</span><span class="n">u_counts</span><span class="p">,</span> <span class="n">v_counts</span><span class="p">,</span> <span class="n">otu_ids</span><span class="p">,</span> <span class="n">tree</span><span class="p">,</span>
<span class="gp">... </span>                      <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">wu</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="go">0.33</span>
</pre></div>
</div>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4a1+.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>