<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Alignments (skbio.alignment) &#8212; scikit-bio 0.5.6 documentation</title>
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/style.css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script src="_static/copybutton.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="skbio.alignment.TabularMSA" href="generated/skbio.alignment.TabularMSA.html" />
    <link rel="prev" title="skbio.sequence.distance.kmer_distance" href="generated/skbio.sequence.distance.kmer_distance.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="_static/bootstrap-sphinx.js "></script>

  </head><body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.5.6</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="io.html">File I/O (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="sequence.html">Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="tree.html">Tree representations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Constructing workflows (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="diversity.html">Diversity calculations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="stats.html">Statistics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Metadata (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.metadata</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="util.html">Utility functionality (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a><ul>
<li><a class="reference internal" href="#data-structures">Data Structures</a></li>
<li><a class="reference internal" href="#optimized-i-e-production-ready-alignment-algorithms">Optimized (i.e., production-ready) Alignment Algorithms</a></li>
<li><a class="reference internal" href="#slow-i-e-educational-purposes-only-alignment-algorithms">Slow (i.e., educational-purposes only) Alignment Algorithms</a></li>
<li><a class="reference internal" href="#general-functionality">General functionality</a></li>
<li><a class="reference internal" href="#data-structure-examples">Data Structure Examples</a></li>
<li><a class="reference internal" href="#alignment-algorithm-examples">Alignment Algorithm Examples</a></li>
<li><a class="reference internal" href="#optimized-alignment-algorithm-examples">Optimized Alignment Algorithm Examples</a></li>
<li><a class="reference internal" href="#slow-alignment-algorithm-examples">Slow Alignment Algorithm Examples</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <span class="target" id="module-skbio.alignment"></span><div class="section" id="alignments-skbio-alignment">
<h1>Alignments (<a class="reference internal" href="#module-skbio.alignment" title="skbio.alignment"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code></a>)<a class="headerlink" href="#alignments-skbio-alignment" title="Permalink to this headline">¶</a></h1>
<p>This module provides functionality for computing and manipulating sequence
alignments. DNA, RNA, and protein sequences can be aligned, as well as
sequences with custom alphabets.</p>
<div class="section" id="data-structures">
<h2>Data Structures<a class="headerlink" href="#data-structures" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TabularMSA</span></code></a>(sequences[, metadata, …])</p></td>
<td><p>Store a multiple sequence alignment in tabular (row/column) form.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="optimized-i-e-production-ready-alignment-algorithms">
<h2>Optimized (i.e., production-ready) Alignment Algorithms<a class="headerlink" href="#optimized-i-e-production-ready-alignment-algorithms" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.StripedSmithWaterman.html#skbio.alignment.StripedSmithWaterman" title="skbio.alignment.StripedSmithWaterman"><code class="xref py py-obj docutils literal notranslate"><span class="pre">StripedSmithWaterman</span></code></a></p></td>
<td><p>Performs a striped (banded) Smith Waterman Alignment.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.AlignmentStructure.html#skbio.alignment.AlignmentStructure" title="skbio.alignment.AlignmentStructure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">AlignmentStructure</span></code></a></p></td>
<td><p>Wraps the result of an alignment c struct so it is accessible to Python</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_ssw.html#skbio.alignment.local_pairwise_align_ssw" title="skbio.alignment.local_pairwise_align_ssw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align_ssw</span></code></a>(sequence1, …)</p></td>
<td><p>Align query and target sequences with Striped Smith-Waterman.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="slow-i-e-educational-purposes-only-alignment-algorithms">
<h2>Slow (i.e., educational-purposes only) Alignment Algorithms<a class="headerlink" href="#slow-i-e-educational-purposes-only-alignment-algorithms" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_nucleotide.html#skbio.alignment.global_pairwise_align_nucleotide" title="skbio.alignment.global_pairwise_align_nucleotide"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_pairwise_align_nucleotide</span></code></a>(seq1, seq2)</p></td>
<td><p>Globally align nucleotide seqs or alignments with Needleman-Wunsch</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_protein.html#skbio.alignment.global_pairwise_align_protein" title="skbio.alignment.global_pairwise_align_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_pairwise_align_protein</span></code></a>(seq1, seq2[, …])</p></td>
<td><p>Globally align pair of protein seqs or alignments with Needleman-Wunsch</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align.html#skbio.alignment.global_pairwise_align" title="skbio.alignment.global_pairwise_align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_pairwise_align</span></code></a>(seq1, seq2, …[, …])</p></td>
<td><p>Globally align a pair of seqs or alignments with Needleman-Wunsch</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_nucleotide.html#skbio.alignment.local_pairwise_align_nucleotide" title="skbio.alignment.local_pairwise_align_nucleotide"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align_nucleotide</span></code></a>(seq1, seq2)</p></td>
<td><p>Locally align exactly two nucleotide seqs with Smith-Waterman</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_protein.html#skbio.alignment.local_pairwise_align_protein" title="skbio.alignment.local_pairwise_align_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align_protein</span></code></a>(seq1, seq2[, …])</p></td>
<td><p>Locally align exactly two protein seqs with Smith-Waterman</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align.html#skbio.alignment.local_pairwise_align" title="skbio.alignment.local_pairwise_align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align</span></code></a>(seq1, seq2, …)</p></td>
<td><p>Locally align exactly two seqs with Smith-Waterman</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="general-functionality">
<h2>General functionality<a class="headerlink" href="#general-functionality" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.make_identity_substitution_matrix.html#skbio.alignment.make_identity_substitution_matrix" title="skbio.alignment.make_identity_substitution_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_identity_substitution_matrix</span></code></a>(…[, …])</p></td>
<td><p>Generate substitution matrix where all matches are scored equally</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="data-structure-examples">
<h2>Data Structure Examples<a class="headerlink" href="#data-structure-examples" title="Permalink to this headline">¶</a></h2>
<p>Load two DNA sequences that have been previously aligned into a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code>
object, using sequence IDs as the MSA’s index:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">TabularMSA</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACC--G-GGTA..&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s2">&quot;seq1&quot;</span><span class="p">}),</span>
<span class="gp">... </span>        <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;TCC--G-GGCA..&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s2">&quot;seq2&quot;</span><span class="p">})]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">minter</span><span class="o">=</span><span class="s1">&#39;id&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">----------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 13</span>
<span class="go">----------------------</span>
<span class="go">ACC--G-GGTA..</span>
<span class="go">TCC--G-GGCA..</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">Index([&#39;seq1&#39;, &#39;seq2&#39;], dtype=&#39;object&#39;)</span>
</pre></div>
</div>
</div>
<div class="section" id="alignment-algorithm-examples">
<h2>Alignment Algorithm Examples<a class="headerlink" href="#alignment-algorithm-examples" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="optimized-alignment-algorithm-examples">
<h2>Optimized Alignment Algorithm Examples<a class="headerlink" href="#optimized-alignment-algorithm-examples" title="Permalink to this headline">¶</a></h2>
<p>Using the convenient <code class="docutils literal notranslate"><span class="pre">local_pairwise_align_ssw</span></code> function:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">local_pairwise_align_ssw</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">start_end_positions</span> <span class="o">=</span> <span class="n">local_pairwise_align_ssw</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACTAAGGCTCTCTACCCCTCTCAGAGA&quot;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACTAAGGCTCCTAACCCCCTTTTCTCAGA&quot;</span><span class="p">)</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">------------------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 30</span>
<span class="go">------------------------------</span>
<span class="go">ACTAAGGCTCTCT-ACCCC----TCTCAGA</span>
<span class="go">ACTAAGGCTC-CTAACCCCCTTTTCTCAGA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">27</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">start_end_positions</span>
<span class="go">[(0, 24), (0, 28)]</span>
</pre></div>
</div>
<p>Using the <code class="docutils literal notranslate"><span class="pre">StripedSmithWaterman</span></code> object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">StripedSmithWaterman</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="n">StripedSmithWaterman</span><span class="p">(</span><span class="s2">&quot;ACTAAGGCTCTCTACCCCTCTCAGAGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">query</span><span class="p">(</span><span class="s2">&quot;AAAAAACTCTCTAAACTCACTAAGGCTCTCTACCCCTCTTCAGAGAAGTCGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">ACTAAGGCTC...</span>
<span class="go">ACTAAGGCTC...</span>
<span class="go">Score: 49</span>
<span class="go">Length: 28</span>
</pre></div>
</div>
<p>Using the <code class="docutils literal notranslate"><span class="pre">StripedSmithWaterman</span></code> object for multiple targets in an efficient
way and finding the aligned sequence representations:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">StripedSmithWaterman</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target_sequences</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>    <span class="s2">&quot;GCTAACTAGGCTCCCTTCTACCCCTCTCAGAGA&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;GCCCAGTAGCTTCCCAATATGAGAGCATCAATTGTAGATCGGGCC&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;TCTATAAGATTCCGCATGCGTTACTTATAAGATGTCTCAACGG&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;TAGAGATTAATTGCCACTGCCAAAATTCTG&quot;</span>
<span class="gp">... </span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query_sequence</span> <span class="o">=</span> <span class="s2">&quot;ACTAAGGCTCTCTACCCCTCTCAGAGA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="n">StripedSmithWaterman</span><span class="p">(</span><span class="n">query_sequence</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">target_sequence</span> <span class="ow">in</span> <span class="n">target_sequences</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">alignment</span> <span class="o">=</span> <span class="n">query</span><span class="p">(</span><span class="n">target_sequence</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">ACTAAGGCTC...</span>
<span class="go">ACT-AGGCTC...</span>
<span class="go">Score: 38</span>
<span class="go">Length: 30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">aligned_query_sequence</span><span class="p">)</span>
<span class="go">ACTAAGGCTC---TCTACCCCTCTCAGAGA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">aligned_target_sequence</span><span class="p">)</span>
<span class="go">ACT-AGGCTCCCTTCTACCCCTCTCAGAGA</span>
</pre></div>
</div>
</div>
<div class="section" id="slow-alignment-algorithm-examples">
<h2>Slow Alignment Algorithm Examples<a class="headerlink" href="#slow-alignment-algorithm-examples" title="Permalink to this headline">¶</a></h2>
<p>scikit-bio also provides pure-Python implementations of Smith-Waterman and
Needleman-Wunsch alignment. These are much slower than the methods described
above, but serve as useful educational examples as they’re simpler to
experiment with. Functions are provided for local and global alignment of
protein and nucleotide sequences. The <code class="docutils literal notranslate"><span class="pre">global*</span></code> and <code class="docutils literal notranslate"><span class="pre">local*</span></code> functions
differ in the underlying algorithm that is applied (<code class="docutils literal notranslate"><span class="pre">global*</span></code> uses Needleman-
Wunsch while <code class="docutils literal notranslate"><span class="pre">local*</span></code> uses Smith-Waterman), and <code class="docutils literal notranslate"><span class="pre">*protein</span></code> and
<code class="docutils literal notranslate"><span class="pre">*nucleotide</span></code> differ in their default scoring of matches, mismatches, and
gaps.</p>
<p>Here we locally align a pair of protein sequences using gap open penalty
of 11 and a gap extend penalty of 1 (in other words, it is much more
costly to open a new gap than extend an existing one).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">Protein</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">local_pairwise_align_protein</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="s2">&quot;HEAGAWGHEE&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="s2">&quot;PAWHEAE&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">start_end_positions</span> <span class="o">=</span> <span class="n">local_pairwise_align_protein</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>This returns an <code class="docutils literal notranslate"><span class="pre">skbio.TabularMSA</span></code> object, the alignment score, and start/end
positions of each aligned sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>
<span class="go">TabularMSA[Protein]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 5</span>
<span class="go">---------------------</span>
<span class="go">AWGHE</span>
<span class="go">AW-HE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">25.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">start_end_positions</span>
<span class="go">[(4, 8), (1, 4)]</span>
</pre></div>
</div>
<p>Similarly, we can perform global alignment of nucleotide sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">global_pairwise_align_nucleotide</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;GCGTGCCTAAGGTATGCAAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACGTGCCTAGGTACGCAAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">start_end_positions</span> <span class="o">=</span> <span class="n">global_pairwise_align_nucleotide</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">----------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 20</span>
<span class="go">----------------------</span>
<span class="go">GCGTGCCTAAGGTATGCAAG</span>
<span class="go">ACGTGCCTA-GGTACGCAAG</span>
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.4.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>