<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>QSeq format (skbio.io.format.qseq) &#8212; scikit-bio 0.5.6 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/style.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Stockholm format (skbio.io.format.stockholm)" href="skbio.io.format.stockholm.html" />
    <link rel="prev" title="PHYLIP multiple sequence alignment format (skbio.io.format.phylip)" href="skbio.io.format.phylip.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../_static/bootstrap-sphinx.js "></script>

  </head><body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.5.6</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../io.html">File I/O (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../metadata.html">Metadata (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.metadata</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">QSeq format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.qseq</span></code>)</a><ul>
<li><a class="reference internal" href="#format-support">Format Support</a></li>
<li><a class="reference internal" href="#format-specification">Format Specification</a></li>
<li><a class="reference internal" href="#format-parameters">Format Parameters</a><ul>
<li><a class="reference internal" href="#generators-only">Generators Only</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <span class="target" id="module-skbio.io.format.qseq"></span><div class="section" id="qseq-format-skbio-io-format-qseq">
<h1>QSeq format (<a class="reference internal" href="#module-skbio.io.format.qseq" title="skbio.io.format.qseq"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.qseq</span></code></a>)<a class="headerlink" href="#qseq-format-skbio-io-format-qseq" title="Permalink to this headline">¶</a></h1>
<p>The QSeq format (<cite>qseq</cite>) is a record-based, plain text output format produced
by some DNA sequencers for storing biological sequence data, quality scores,
per-sequence filtering information, and run-specific metadata.</p>
<div class="section" id="format-support">
<h2>Format Support<a class="headerlink" href="#format-support" title="Permalink to this headline">¶</a></h2>
<p><strong>Has Sniffer: Yes</strong></p>
<table class="docutils align-default">
<colgroup>
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 84%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Reader</p></th>
<th class="head"><p>Writer</p></th>
<th class="head"><p>Object Class</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>No</p></td>
<td><p>generator of <a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a> objects</p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>No</p></td>
<td><p><a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>No</p></td>
<td><p><a class="reference internal" href="skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.DNA</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>No</p></td>
<td><p><a class="reference internal" href="skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.RNA</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>No</p></td>
<td><p><a class="reference internal" href="skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Protein</span></code></a></p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="format-specification">
<h2>Format Specification<a class="headerlink" href="#format-specification" title="Permalink to this headline">¶</a></h2>
<p>A QSeq file is composed of single-line records, delimited by tabs. There are
11 fields in a record:</p>
<ul class="simple">
<li><p>Machine name</p></li>
<li><p>Run number</p></li>
<li><p>Lane number (positive int)</p></li>
<li><p>Tile number (positive int)</p></li>
<li><p>X coordinate (integer)</p></li>
<li><p>Y coordinate (integer)</p></li>
<li><p>Index</p></li>
<li><p>Read number (1-3)</p></li>
<li><p>Sequence data (typically IUPAC characters)</p></li>
<li><p>Quality scores (quality scores encoded as printable ASCII)</p></li>
<li><p>Filter boolean (1 if sequence has passed CASAVA’s filter, 0 otherwise)</p></li>
</ul>
<p>For more details please refer to the CASAVA documentation <a class="footnote-reference brackets" href="#id2" id="id1">1</a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When a QSeq file is read into a scikit-bio object, the object’s
<cite>metadata</cite> attribute is automatically populated with data corresponding
to the names above.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><cite>lowercase</cite> functionality is supported when reading QSeq files.
Refer to specific object constructor documentation for details.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>scikit-bio allows for the filter field to be ommitted, but it is not
clear if this is part of the original format specification.</p>
</div>
</div>
<div class="section" id="format-parameters">
<h2>Format Parameters<a class="headerlink" href="#format-parameters" title="Permalink to this headline">¶</a></h2>
<p>The following parameters are the same as in FASTQ format
(<a class="reference internal" href="skbio.io.format.fastq.html#module-skbio.io.format.fastq" title="skbio.io.format.fastq"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fastq</span></code></a>):</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">variant</span></code>: see <code class="docutils literal notranslate"><span class="pre">variant</span></code> parameter in FASTQ format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">phred_offset</span></code>: see <code class="docutils literal notranslate"><span class="pre">phred_offset</span></code> parameter in FASTQ format</p></li>
</ul>
<p>The following additional parameters are the same as in FASTA format
(<a class="reference internal" href="skbio.io.format.fasta.html#module-skbio.io.format.fasta" title="skbio.io.format.fasta"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fasta</span></code></a>):</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">constructor</span></code>: see <code class="docutils literal notranslate"><span class="pre">constructor</span></code> parameter in FASTA format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">seq_num</span></code>: see <code class="docutils literal notranslate"><span class="pre">seq_num</span></code> parameter in FASTA format</p></li>
</ul>
<div class="section" id="generators-only">
<h3>Generators Only<a class="headerlink" href="#generators-only" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">filter</span></code>: If <cite>True</cite>, excludes sequences that did not pass filtering
(i.e., filter field is 0). Default is <cite>True</cite>.</p></li>
</ul>
<p class="rubric">Examples</p>
<p>Suppose we have the following QSeq file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">illumina</span>    <span class="mi">1</span>       <span class="mi">3</span>       <span class="mi">34</span>      <span class="o">-</span><span class="mi">30</span>     <span class="mi">30</span>      <span class="mi">0</span>       <span class="mi">1</span>       <span class="n">ACG</span><span class="o">....</span><span class="n">ACGTAC</span>   <span class="n">ruBBBBrBCEFGH</span>   <span class="mi">1</span>
<span class="n">illumina</span>    <span class="mi">1</span>       <span class="mi">3</span>       <span class="mi">34</span>      <span class="mi">30</span>      <span class="o">-</span><span class="mi">30</span>     <span class="mi">0</span>       <span class="mi">1</span>       <span class="n">CGGGCATTGCA</span>     <span class="n">CGGGCasdGCA</span>     <span class="mi">0</span>
<span class="n">illumina</span>    <span class="mi">1</span>       <span class="mi">3</span>       <span class="mi">35</span>      <span class="o">-</span><span class="mi">30</span>     <span class="mi">30</span>      <span class="mi">0</span>       <span class="mi">2</span>       <span class="n">ACGTA</span><span class="o">.</span><span class="n">AATAAAC</span>   <span class="n">geTaAafhwqAAf</span>   <span class="mi">1</span>
<span class="n">illumina</span>    <span class="mi">1</span>       <span class="mi">3</span>       <span class="mi">35</span>      <span class="mi">30</span>      <span class="o">-</span><span class="mi">30</span>     <span class="mi">0</span>       <span class="mi">3</span>       <span class="n">CATTTAGGA</span><span class="o">.</span><span class="n">TGCA</span>  <span class="n">tjflkAFnkKghvM</span>  <span class="mi">0</span>
</pre></div>
</div>
<p>Let’s define this file in-memory as a <code class="docutils literal notranslate"><span class="pre">StringIO</span></code>, though this could be a real
file path, file handle, or anything that’s supported by scikit-bio’s I/O
registry in practice:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fs</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>    <span class="s1">&#39;illumina</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">3</span><span class="se">\t</span><span class="s1">34</span><span class="se">\t</span><span class="s1">-30</span><span class="se">\t</span><span class="s1">30</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">ACG....ACGTAC</span><span class="se">\t</span><span class="s1">ruBBBBrBCEFGH</span><span class="se">\t</span><span class="s1">1&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;illumina</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">3</span><span class="se">\t</span><span class="s1">34</span><span class="se">\t</span><span class="s1">30</span><span class="se">\t</span><span class="s1">-30</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">CGGGCATTGCA</span><span class="se">\t</span><span class="s1">CGGGCasdGCA</span><span class="se">\t</span><span class="s1">0&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;illumina</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">3</span><span class="se">\t</span><span class="s1">35</span><span class="se">\t</span><span class="s1">-30</span><span class="se">\t</span><span class="s1">30</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">2</span><span class="se">\t</span><span class="s1">ACGTA.AATAAAC</span><span class="se">\t</span><span class="s1">geTaAafhwqAAf</span><span class="se">\t</span><span class="s1">1&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;illumina</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">3</span><span class="se">\t</span><span class="s1">35</span><span class="se">\t</span><span class="s1">30</span><span class="se">\t</span><span class="s1">-30</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">3</span><span class="se">\t</span><span class="s1">CATTTAGGA.TGCA</span><span class="se">\t</span><span class="s1">tjflkAFnkKghvM</span><span class="se">\t</span><span class="s1">0&#39;</span>
<span class="gp">... </span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span>
</pre></div>
</div>
<p>To iterate over the sequences using the generator reader, we run:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">skbio.io</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">skbio</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;qseq&#39;</span><span class="p">,</span> <span class="n">variant</span><span class="o">=</span><span class="s1">&#39;illumina1.3&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">seq</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="go">Sequence</span>
<span class="go">--------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;id&#39;: &#39;illumina_1:3:34:-30:30#0/1&#39;</span>
<span class="go">    &#39;index&#39;: 0</span>
<span class="go">    &#39;lane_number&#39;: 3</span>
<span class="go">    &#39;machine_name&#39;: &#39;illumina&#39;</span>
<span class="go">    &#39;read_number&#39;: 1</span>
<span class="go">    &#39;run_number&#39;: 1</span>
<span class="go">    &#39;tile_number&#39;: 34</span>
<span class="go">    &#39;x&#39;: -30</span>
<span class="go">    &#39;y&#39;: 30</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: uint8&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 13</span>
<span class="go">--------------------------------------</span>
<span class="go">0 ACG....ACG TAC</span>

<span class="go">Sequence</span>
<span class="go">--------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;id&#39;: &#39;illumina_1:3:35:-30:30#0/2&#39;</span>
<span class="go">    &#39;index&#39;: 0</span>
<span class="go">    &#39;lane_number&#39;: 3</span>
<span class="go">    &#39;machine_name&#39;: &#39;illumina&#39;</span>
<span class="go">    &#39;read_number&#39;: 2</span>
<span class="go">    &#39;run_number&#39;: 1</span>
<span class="go">    &#39;tile_number&#39;: 35</span>
<span class="go">    &#39;x&#39;: -30</span>
<span class="go">    &#39;y&#39;: 30</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: uint8&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 13</span>
<span class="go">--------------------------------------</span>
<span class="go">0 ACGTA.AATA AAC</span>
</pre></div>
</div>
<p>Note that only two sequences were loaded because the QSeq reader filters out
sequences whose filter field is 0 (unless <code class="docutils literal notranslate"><span class="pre">filter=False</span></code> is supplied).</p>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id2"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p><a class="reference external" href="http://biowulf.nih.gov/apps/CASAVA_UG_15011196B.pdf">http://biowulf.nih.gov/apps/CASAVA_UG_15011196B.pdf</a></p>
</dd>
</dl>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.4.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>