<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>BLAST+6 format (skbio.io.format.blast6) &#8212; scikit-bio 0.5.6 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/style.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="BLAST+7 format (skbio.io.format.blast7)" href="skbio.io.format.blast7.html" />
    <link rel="prev" title="File I/O (skbio.io)" href="../io.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../_static/bootstrap-sphinx.js "></script>

  </head><body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.5.6</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../io.html">File I/O (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../metadata.html">Metadata (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.metadata</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">BLAST+6 format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.blast6</span></code>)</a><ul>
<li><a class="reference internal" href="#format-support">Format Support</a></li>
<li><a class="reference internal" href="#format-specification">Format Specification</a><ul>
<li><a class="reference internal" href="#blast-column-types">BLAST Column Types</a></li>
</ul>
</li>
<li><a class="reference internal" href="#format-parameters">Format Parameters</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <span class="target" id="module-skbio.io.format.blast6"></span><div class="section" id="blast-6-format-skbio-io-format-blast6">
<h1>BLAST+6 format (<a class="reference internal" href="#module-skbio.io.format.blast6" title="skbio.io.format.blast6"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.blast6</span></code></a>)<a class="headerlink" href="#blast-6-format-skbio-io-format-blast6" title="Permalink to this headline">¶</a></h1>
<p>The BLAST+6 format (<code class="docutils literal notranslate"><span class="pre">blast+6</span></code>) stores the results of a BLAST <a class="footnote-reference brackets" href="#id4" id="id1">1</a> database
search. The results are stored in a simple tabular format with no column
headers. Values are separated by the tab character.</p>
<p>An example BLAST+6-formatted file comparing two protein sequences, taken
from <a class="footnote-reference brackets" href="#id5" id="id2">2</a> (tab characters represented by <code class="docutils literal notranslate"><span class="pre">&lt;tab&gt;</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">moaC</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="n">gi</span><span class="o">|</span><span class="mi">15800534</span><span class="o">|</span><span class="n">ref</span><span class="o">|</span><span class="n">NP_286546</span><span class="o">.</span><span class="mi">1</span><span class="o">|&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mf">100.00</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">161</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">0</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">0</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">1</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">161</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">1</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">161</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mf">3e-114</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">330</span>
<span class="n">moaC</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="n">gi</span><span class="o">|</span><span class="mi">170768970</span><span class="o">|</span><span class="n">ref</span><span class="o">|</span><span class="n">ZP_02903423</span><span class="o">.</span><span class="mi">1</span><span class="o">|&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mf">99.38</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">161</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">1</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">0</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">1</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">161</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">1</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">161</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mf">9e-114</span><span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span><span class="mi">329</span>
</pre></div>
</div>
<div class="section" id="format-support">
<h2>Format Support<a class="headerlink" href="#format-support" title="Permalink to this headline">¶</a></h2>
<p><strong>Has Sniffer: No</strong></p>
<p><strong>State: Experimental as of 0.4.1.</strong></p>
<table class="docutils align-default">
<colgroup>
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 84%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Reader</p></th>
<th class="head"><p>Writer</p></th>
<th class="head"><p>Object Class</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>No</p></td>
<td><p><code class="xref py py-mod docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="format-specification">
<h2>Format Specification<a class="headerlink" href="#format-specification" title="Permalink to this headline">¶</a></h2>
<p>BLAST+6 format is a tabular text-based format produced by both BLAST+ output
format 6 (<code class="docutils literal notranslate"><span class="pre">-outfmt</span> <span class="pre">6</span></code>) and legacy BLAST output format 8 (<code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">8</span></code>). It is
tab-separated and has no column headers. With BLAST+, users can specify the
columns that are present in their BLAST output file by specifying column names
(e.g., <code class="docutils literal notranslate"><span class="pre">-outfmt</span> <span class="pre">&quot;6</span> <span class="pre">qseqid</span> <span class="pre">sseqid</span> <span class="pre">bitscore</span> <span class="pre">qstart</span> <span class="pre">sstart&quot;</span></code>), if the default
columns output by BLAST are not desired.</p>
<div class="section" id="blast-column-types">
<h3>BLAST Column Types<a class="headerlink" href="#blast-column-types" title="Permalink to this headline">¶</a></h3>
<p>The following column types are output by BLAST and supported by scikit-bio.
This information is taken from <a class="footnote-reference brackets" href="#id6" id="id3">3</a>.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 21%" />
<col style="width: 69%" />
<col style="width: 10%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>qseqid</p></td>
<td><p>Query Seq-id</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>qgi</p></td>
<td><p>Query GI</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>qacc</p></td>
<td><p>Query accesion</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>qaccver</p></td>
<td><p>Query accesion.version</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>qlen</p></td>
<td><p>Query sequence length</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>sseqid</p></td>
<td><p>Subject Seq-id</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>sallseqid</p></td>
<td><p>All subject Seq-id(s), separated by
a ‘;’</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>sgi</p></td>
<td><p>Subject GI</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>sallgi</p></td>
<td><p>All subject GIs</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>sacc</p></td>
<td><p>Subject accesion</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>saccver</p></td>
<td><p>Subject accesion.version</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>sallacc</p></td>
<td><p>All subject accesions</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>slen</p></td>
<td><p>Subject sequence length</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>qstart</p></td>
<td><p>Start of alignment in query</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>qend</p></td>
<td><p>End of alignment in query</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>sstart</p></td>
<td><p>Start of alignment in subject</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>send</p></td>
<td><p>End of alignment in subject</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>qseq</p></td>
<td><p>Aligned part of query sequence</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>sseq</p></td>
<td><p>Aligned part of subject sequence</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>evalue</p></td>
<td><p>Expect value</p></td>
<td><p>float</p></td>
</tr>
<tr class="row-even"><td><p>bitscore</p></td>
<td><p>Bit score</p></td>
<td><p>float</p></td>
</tr>
<tr class="row-odd"><td><p>score</p></td>
<td><p>Raw score</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>length</p></td>
<td><p>Alignment length</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>pident</p></td>
<td><p>Percent of identical matches</p></td>
<td><p>float</p></td>
</tr>
<tr class="row-even"><td><p>nident</p></td>
<td><p>Number of identical matches</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>mismatch</p></td>
<td><p>Number of mismatches</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>positive</p></td>
<td><p>Number of positive-scoring matches</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>gapopen</p></td>
<td><p>Number of gap openings</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>gaps</p></td>
<td><p>Total number of gaps</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>ppos</p></td>
<td><p>Percentage of positive-scoring
matches</p></td>
<td><p>float</p></td>
</tr>
<tr class="row-even"><td><p>frames</p></td>
<td><p>Query and subject frames separated
by a ‘/’</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>qframe</p></td>
<td><p>Query frame</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>sframe</p></td>
<td><p>Subject frame</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>btop</p></td>
<td><p>Blast traceback operations (BTOP)</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-even"><td><p>staxids</p></td>
<td><p>Unique Subject Taxonomy ID(s),
separated by a ‘;’ (in numerical
order).</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>sscinames</p></td>
<td><p>Unique Subject Scientific Name(s),
separated by a ‘;’</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>scomnames</p></td>
<td><p>Unique Subject Common Name(s),
separated by a ‘;’</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>sblastnames</p></td>
<td><p>unique Subject Blast Name(s),
separated by a ‘;’ (in alphabetical
order)</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>sskingdoms</p></td>
<td><p>unique Subject Super Kingdom(s),
separated by a ‘;’ (in alphabetical
order)</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>stitle</p></td>
<td><p>Subject Title</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>sstrand</p></td>
<td><p>Subject Strand</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-odd"><td><p>salltitles</p></td>
<td><p>All Subject Title(s), separated by
a ‘&lt;&gt;’</p></td>
<td><p>str</p></td>
</tr>
<tr class="row-even"><td><p>qcovs</p></td>
<td><p>Query Coverage Per Subject</p></td>
<td><p>int</p></td>
</tr>
<tr class="row-odd"><td><p>qcovhsp</p></td>
<td><p>Query Coverage Per HSP</p></td>
<td><p>int</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When a BLAST+6-formatted file contains <code class="docutils literal notranslate"><span class="pre">N/A</span></code> values, scikit-bio
will convert these values into <code class="docutils literal notranslate"><span class="pre">np.nan</span></code>, matching pandas’ convention for
representing missing data.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>scikit-bio stores columns of type <code class="docutils literal notranslate"><span class="pre">int</span></code> as type <code class="docutils literal notranslate"><span class="pre">float</span></code> in the
returned <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>. This is necessary in order to allow <code class="docutils literal notranslate"><span class="pre">N/A</span></code>
values in integer columns (this is currently a limitation of pandas).</p>
</div>
</div>
</div>
<div class="section" id="format-parameters">
<h2>Format Parameters<a class="headerlink" href="#format-parameters" title="Permalink to this headline">¶</a></h2>
<p>The following format parameters are available in <code class="docutils literal notranslate"><span class="pre">blast+6</span></code> format:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">default_columns</span></code>: <code class="docutils literal notranslate"><span class="pre">False</span></code> by default. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, will use the default
columns output by BLAST, which are qseqid, sseqid, pident, length, mismatch,
gapopen, qstart, qend, sstart, send, evalue, and bitscore.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>When reading legacy BLAST files, you must pass
<code class="docutils literal notranslate"><span class="pre">default_columns=True</span></code> because legacy BLAST does not allow users to
specify which columns are present in the output file.</p>
</div>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">columns</span></code>: <code class="docutils literal notranslate"><span class="pre">None</span></code> by default. If provided, must be a list of column names
in the order they will appear in the file.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Either <code class="docutils literal notranslate"><span class="pre">default_columns</span></code> or <code class="docutils literal notranslate"><span class="pre">columns</span></code> must be provided, as
<code class="docutils literal notranslate"><span class="pre">blast+6</span></code> does not contain column headers.</p>
</div>
<p class="rubric">Examples</p>
<p>Suppose we have a <code class="docutils literal notranslate"><span class="pre">blast+6</span></code> file with default columns:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">skbio.io</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fs</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>    <span class="s1">&#39;moaC</span><span class="se">\t</span><span class="s1">gi|15800534|ref|NP_286546.1|</span><span class="se">\t</span><span class="s1">100.00</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">3e-114</span><span class="se">\t</span><span class="s1">330&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;moaC</span><span class="se">\t</span><span class="s1">gi|170768970|ref|ZP_02903423.1|</span><span class="se">\t</span><span class="s1">99.38</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">9e-114</span><span class="se">\t</span><span class="s1">329&#39;</span>
<span class="gp">... </span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span>
</pre></div>
</div>
<p>Read the file into a <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> and specify that default columns should
be used:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">skbio</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;blast+6&quot;</span><span class="p">,</span> <span class="n">into</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
<span class="gp">... </span>                   <span class="n">default_columns</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> 
<span class="go">  qseqid                           sseqid  pident  length  mismatch  gapopen \</span>
<span class="go">0   moaC     gi|15800534|ref|NP_286546.1|  100.00   161.0       0.0      0.0</span>
<span class="go">1   moaC  gi|170768970|ref|ZP_02903423.1|   99.38   161.0       1.0      0.0</span>

<span class="go">   qstart   qend  sstart   send         evalue  bitscore</span>
<span class="go">0     1.0  161.0     1.0  161.0  3.000000e-114     330.0</span>
<span class="go">1     1.0  161.0     1.0  161.0  9.000000e-114     329.0</span>
</pre></div>
</div>
<p>Suppose we have a <code class="docutils literal notranslate"><span class="pre">blast+6</span></code> file with user-supplied (non-default) columns:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">skbio.io</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fs</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>    <span class="s1">&#39;moaC</span><span class="se">\t</span><span class="s1">100.00</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">330</span><span class="se">\t</span><span class="s1">1&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;moaC</span><span class="se">\t</span><span class="s1">99.38</span><span class="se">\t</span><span class="s1">1</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">0</span><span class="se">\t</span><span class="s1">161</span><span class="se">\t</span><span class="s1">329</span><span class="se">\t</span><span class="s1">1&#39;</span>
<span class="gp">... </span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span>
</pre></div>
</div>
<p>Read the file into a <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> and specify which columns are present
in the file:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">skbio</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;blast+6&quot;</span><span class="p">,</span> <span class="n">into</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
<span class="gp">... </span>                   <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;qseqid&#39;</span><span class="p">,</span> <span class="s1">&#39;pident&#39;</span><span class="p">,</span> <span class="s1">&#39;mismatch&#39;</span><span class="p">,</span> <span class="s1">&#39;length&#39;</span><span class="p">,</span>
<span class="gp">... </span>                            <span class="s1">&#39;gapopen&#39;</span><span class="p">,</span> <span class="s1">&#39;qend&#39;</span><span class="p">,</span> <span class="s1">&#39;bitscore&#39;</span><span class="p">,</span> <span class="s1">&#39;sstart&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> 
<span class="go">  qseqid  pident  mismatch  length  gapopen   qend  bitscore  sstart</span>
<span class="go">0   moaC  100.00       0.0   161.0      0.0  161.0     330.0     1.0</span>
<span class="go">1   moaC   99.38       1.0   161.0      0.0  161.0     329.0     1.0</span>
</pre></div>
</div>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id4"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Altschul, S.F., Gish, W., Miller, W., Myers, E.W. &amp; Lipman, D.J. (1990)
“Basic local alignment search tool.” J. Mol. Biol. 215:403-410.</p>
</dd>
<dt class="label" id="id5"><span class="brackets"><a class="fn-backref" href="#id2">2</a></span></dt>
<dd><p><a class="reference external" href="http://blastedbio.blogspot.com/2014/11/column-headers-in-blast-tabular-and-csv.html">http://blastedbio.blogspot.com/2014/11/column-headers-in-blast-tabular-and-csv.html</a></p>
</dd>
<dt class="label" id="id6"><span class="brackets"><a class="fn-backref" href="#id3">3</a></span></dt>
<dd><p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK279675/">http://www.ncbi.nlm.nih.gov/books/NBK279675/</a></p>
</dd>
</dl>
</div>
</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.4.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>