<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>FASTQ format (skbio.io.format.fastq) &#8212; scikit-bio 0.5.6 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/style.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="GenBank format (skbio.io.format.genbank)" href="skbio.io.format.genbank.html" />
    <link rel="prev" title="FASTA/QUAL format (skbio.io.format.fasta)" href="skbio.io.format.fasta.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../_static/bootstrap-sphinx.js "></script>

  </head><body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.5.6</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../io.html">File I/O (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../metadata.html">Metadata (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.metadata</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">FASTQ format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fastq</span></code>)</a><ul>
<li><a class="reference internal" href="#format-support">Format Support</a></li>
<li><a class="reference internal" href="#format-specification">Format Specification</a><ul>
<li><a class="reference internal" href="#quality-score-variants">Quality Score Variants</a></li>
</ul>
</li>
<li><a class="reference internal" href="#format-parameters">Format Parameters</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <span class="target" id="module-skbio.io.format.fastq"></span><div class="section" id="fastq-format-skbio-io-format-fastq">
<h1>FASTQ format (<a class="reference internal" href="#module-skbio.io.format.fastq" title="skbio.io.format.fastq"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fastq</span></code></a>)<a class="headerlink" href="#fastq-format-skbio-io-format-fastq" title="Permalink to this headline">¶</a></h1>
<p>The FASTQ file format (<code class="docutils literal notranslate"><span class="pre">fastq</span></code>) stores biological (e.g., nucleotide)
sequences and their quality scores in a simple plain text format that is both
human-readable and easy to parse. The file format was invented by Jim Mullikin
at the Wellcome Trust Sanger Institute but wasn’t given a formal definition,
though it has informally become a standard file format for storing
high-throughput sequence data. More information about the format and its
variants can be found in <a class="footnote-reference brackets" href="#id5" id="id1">1</a> and <a class="footnote-reference brackets" href="#id6" id="id2">2</a>.</p>
<p>Conceptually, a FASTQ file is similar to paired FASTA and QUAL files in that it
stores both biological sequences and their quality scores. FASTQ differs from
FASTA/QUAL because the quality scores are stored in the same file as the
biological sequence data.</p>
<p>An example FASTQ-formatted file containing two DNA sequences and their quality
scores:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>@seq1 description 1
AACACCAAACTTCTCCACCACGTGAGCTACAAAAG
+
````Y^T]`]c^cabcacc`^Lb^ccYT\T\Y\WF
@seq2 description 2
TATGTATATATAACATATACATATATACATACATA
+
]KZ[PY]_[YY^```ac^\\`bT``c`\aT``bbb
</pre></div>
</div>
<div class="section" id="format-support">
<h2>Format Support<a class="headerlink" href="#format-support" title="Permalink to this headline">¶</a></h2>
<p><strong>Has Sniffer: Yes</strong></p>
<table class="docutils align-default">
<colgroup>
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 84%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Reader</p></th>
<th class="head"><p>Writer</p></th>
<th class="head"><p>Object Class</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p>generator of <a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a> objects</p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment.TabularMSA</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.DNA</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.RNA</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Protein</span></code></a></p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="format-specification">
<h2>Format Specification<a class="headerlink" href="#format-specification" title="Permalink to this headline">¶</a></h2>
<p>A FASTQ file contains one or more biological sequences and their corresponding
quality scores stored sequentially as <em>records</em>. Each <em>record</em> consists of four
sections:</p>
<ol class="arabic simple">
<li><p>Sequence header line consisting of a sequence identifier (ID) and
description (both optional)</p></li>
<li><p>Biological sequence data (typically stored using the standard IUPAC
lexicon), optionally split over multiple lines</p></li>
<li><p>Quality header line separating sequence data from quality scores (optionally
repeating the ID and description from the sequence header line)</p></li>
<li><p>Quality scores as printable ASCII characters, optionally split over multiple
lines. Decoding of quality scores will depend on the specified FASTQ variant
(see below for more details)</p></li>
</ol>
<p>For the complete FASTQ format specification, see <a class="footnote-reference brackets" href="#id5" id="id3">1</a>. scikit-bio’s FASTQ
implementation follows the format specification described in this excellent
publication, including validating the implementation against the FASTQ example
files provided in the publication’s supplementary data.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>IDs and descriptions will be parsed from sequence header lines in
exactly the same way as FASTA headers (<a class="reference internal" href="skbio.io.format.fasta.html#module-skbio.io.format.fasta" title="skbio.io.format.fasta"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fasta</span></code></a>). IDs,
descriptions, and quality scores are also stored on, and written from,
sequence objects in the same way as with FASTA.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Blank or whitespace-only lines are only allowed at the beginning of
the file, between FASTQ records, or at the end of the file. A blank or
whitespace-only line after the header line, within the sequence, or within
quality scores will raise an error.</p>
<p>scikit-bio will ignore leading and trailing whitespace characters on each
line while reading.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Validation may be performed depending on the type of object the data
is being read into. This behavior matches that of FASTA files.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>scikit-bio will write FASTQ files in a normalized format, with each
record section on a single line. Thus, each record will be composed of
<em>exactly</em> four lines. The quality header line won’t have the sequence ID and
description repeated.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><cite>lowercase</cite> functionality is supported the same as with FASTA.</p>
</div>
<div class="section" id="quality-score-variants">
<h3>Quality Score Variants<a class="headerlink" href="#quality-score-variants" title="Permalink to this headline">¶</a></h3>
<p>FASTQ associates quality scores with sequence data, with each quality score
encoded as a single printable ASCII character. In scikit-bio, all quality
scores are decoded as Phred quality scores. This is the most common quality
score metric, though there are others (e.g., Solexa quality scores).
Unfortunately, different sequencers have different ways of encoding quality
scores as ASCII characters, notably Sanger and Illumina. Below is a table
highlighting the different encoding variants supported by scikit-bio, as well
as listing the equivalent variant names used in the Open Bioinformatics
Foundation (OBF) <a class="footnote-reference brackets" href="#id7" id="id4">3</a> projects (e.g., Biopython, BioPerl, etc.).</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 15%" />
<col style="width: 12%" />
<col style="width: 5%" />
<col style="width: 11%" />
<col style="width: 56%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Variant</p></th>
<th class="head"><p>ASCII
Range</p></th>
<th class="head"><p>Offset</p></th>
<th class="head"><p>Quality
Range</p></th>
<th class="head"><p>Notes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sanger</p></td>
<td><p>33 to 126</p></td>
<td><p>33</p></td>
<td><p>0 to 93</p></td>
<td><p>Equivalent to OBF’s fastq-sanger.</p></td>
</tr>
<tr class="row-odd"><td><p>illumina1.3</p></td>
<td><p>64 to 126</p></td>
<td><p>64</p></td>
<td><p>0 to 62</p></td>
<td><p>Equivalent to OBF’s fastq-illumina. Use
this if your data was generated using
Illumina 1.3-1.7 software.</p></td>
</tr>
<tr class="row-even"><td><p>illumina1.8</p></td>
<td><p>33 to 95</p></td>
<td><p>33</p></td>
<td><p>0 to 62</p></td>
<td><p>Equivalent to sanger but with 0 to 62
quality score range check. Use this if
your data was generated using Illumina
1.8 software or later.</p></td>
</tr>
<tr class="row-odd"><td><p>solexa</p></td>
<td><p>59 to 126</p></td>
<td><p>64</p></td>
<td><p>-5 to 62</p></td>
<td><p>Not currently implemented.</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When writing, Phred quality scores will be truncated to the maximum
value in the variant’s range and a warning will be issued. This is
consistent with the OBF projects.</p>
<p>When reading, an error will be raised if a decoded quality score is outside
the variant’s range.</p>
</div>
</div>
</div>
<div class="section" id="format-parameters">
<h2>Format Parameters<a class="headerlink" href="#format-parameters" title="Permalink to this headline">¶</a></h2>
<p>The following parameters are available to all FASTQ format readers and writers:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">variant</span></code>: A string indicating the quality score variant used to
decode/encode Phred quality scores. Must be one of <code class="docutils literal notranslate"><span class="pre">sanger</span></code>,
<code class="docutils literal notranslate"><span class="pre">illumina1.3</span></code>, <code class="docutils literal notranslate"><span class="pre">illumina1.8</span></code>, or <code class="docutils literal notranslate"><span class="pre">solexa</span></code>. This parameter is preferred
over <code class="docutils literal notranslate"><span class="pre">phred_offset</span></code> because additional quality score range checks and
conversions can be performed. It is also more explicit.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">phred_offset</span></code>: An integer indicating the ASCII code offset used to
decode/encode Phred quality scores. Must be in the range <code class="docutils literal notranslate"><span class="pre">[33,</span> <span class="pre">126]</span></code>. All
decoded scores will be assumed to be Phred scores (i.e., no additional
conversions are performed). Prefer using <code class="docutils literal notranslate"><span class="pre">variant</span></code> over this parameter
whenever possible.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You must provide <code class="docutils literal notranslate"><span class="pre">variant</span></code> or <code class="docutils literal notranslate"><span class="pre">phred_offset</span></code> when reading or
writing a FASTQ file. <code class="docutils literal notranslate"><span class="pre">variant</span></code> and <code class="docutils literal notranslate"><span class="pre">phred_offset</span></code> cannot both be
provided at the same time.</p>
</div>
<p>The following additional parameters are the same as in FASTA format
(<a class="reference internal" href="skbio.io.format.fasta.html#module-skbio.io.format.fasta" title="skbio.io.format.fasta"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fasta</span></code></a>):</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">constructor</span></code>: see <code class="docutils literal notranslate"><span class="pre">constructor</span></code> parameter in FASTA format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">seq_num</span></code>: see <code class="docutils literal notranslate"><span class="pre">seq_num</span></code> parameter in FASTA format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">id_whitespace_replacement</span></code>: see <code class="docutils literal notranslate"><span class="pre">id_whitespace_replacement</span></code> parameter in
FASTA format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">description_newline_replacement</span></code>: see <code class="docutils literal notranslate"><span class="pre">description_newline_replacement</span></code>
parameter in FASTA format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lowercase</span></code>: see <code class="docutils literal notranslate"><span class="pre">lowercase</span></code> parameter in FASTA format</p></li>
</ul>
<p class="rubric">Examples</p>
<p>Suppose we have the following FASTQ file with two DNA sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@seq1</span> <span class="n">description</span> <span class="mi">1</span>
<span class="n">AACACCAAACTTCTCCACC</span>
<span class="n">ACGTGAGCTACAAAAG</span>
<span class="o">+</span><span class="n">seq1</span> <span class="n">description</span> <span class="mi">1</span>
<span class="sd">&#39;&#39;&#39;&#39;Y^T]&#39;]C^CABCACC</span>
<span class="sd">`^LB^CCYT\T\Y\WF</span>
<span class="sd">@seq2 description 2</span>
<span class="sd">TATGTATATATAACATATACATATATACATACATA</span>
<span class="sd">+</span>
<span class="sd">]KZ[PY]_[YY^&#39;&#39;&#39;</span><span class="n">AC</span><span class="o">^</span>\\<span class="s1">&#39;BT&#39;&#39;C&#39;</span>\<span class="n">AT</span><span class="s1">&#39;&#39;</span><span class="n">BBB</span>
</pre></div>
</div>
<p>Note that the first sequence and its quality scores are split across multiple
lines, while the second sequence and its quality scores are each on a single
line. Also note that the first sequence has a duplicate ID and description on
the quality header line, while the second sequence does not.</p>
<p>Let’s define this file in-memory as a <code class="docutils literal notranslate"><span class="pre">StringIO</span></code>, though this could be a real
file path, file handle, or anything that’s supported by scikit-bio’s I/O
registry in practice:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fs</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;@seq1 description 1&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;AACACCAAACTTCTCCACC&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;ACGTGAGCTACAAAAG&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;+seq1 description 1&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;&#39;&#39;&#39;&#39;Y^T]&#39;]C^CABCACC&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;&#39;^LB^CCYT\T\Y\WF&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;@seq2 description 2&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;TATGTATATATAACATATACATATATACATACATA&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;]KZ[PY]_[YY^&#39;&#39;&#39;AC^</span><span class="se">\\</span><span class="s2">&#39;BT&#39;&#39;C&#39;\AT&#39;&#39;BBB&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span>
</pre></div>
</div>
<p>To load the sequences into a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code>, we run:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">TabularMSA</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">constructor</span><span class="o">=</span><span class="n">DNA</span><span class="p">,</span> <span class="n">variant</span><span class="o">=</span><span class="s1">&#39;sanger&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">-----------------------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 35</span>
<span class="go">-----------------------------------</span>
<span class="go">AACACCAAACTTCTCCACCACGTGAGCTACAAAAG</span>
<span class="go">TATGTATATATAACATATACATATATACATACATA</span>
</pre></div>
</div>
<p>Note that quality scores are decoded from Sanger. To load the second sequence
as <code class="docutils literal notranslate"><span class="pre">DNA</span></code>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span> <span class="c1"># reload the StringIO to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">variant</span><span class="o">=</span><span class="s1">&#39;sanger&#39;</span><span class="p">,</span> <span class="n">seq_num</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">DNA</span>
<span class="go">----------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;description 2&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq2&#39;</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: uint8&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 35</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 14.29%</span>
<span class="go">----------------------------------------</span>
<span class="go">0 TATGTATATA TAACATATAC ATATATACAT ACATA</span>
</pre></div>
</div>
<p>To write our <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> to a FASTQ file with quality scores encoded using
the <code class="docutils literal notranslate"><span class="pre">illumina1.3</span></code> variant:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">new_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_fh</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fastq&#39;</span><span class="p">,</span> <span class="n">variant</span><span class="o">=</span><span class="s1">&#39;illumina1.3&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">@seq1 description 1</span>
<span class="go">AACACCAAACTTCTCCACCACGTGAGCTACAAAAG</span>
<span class="go">+</span>
<span class="go">FFFFx}s|F|b}b`ab`bbF}ka}bbxs{s{x{ve</span>
<span class="go">@seq2 description 2</span>
<span class="go">TATGTATATATAACATATACATATATACATACATA</span>
<span class="go">+</span>
<span class="go">|jyzox|~zxx}FFF`b}{{FasFFbF{`sFFaaa</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">new_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Note that the file has been written in normalized format: sequence and quality
scores each only occur on a single line and the sequence header line is
not repeated in the quality header line. Note also that the quality scores are
different because they have been encoded using a different variant.</p>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id5"><span class="brackets">1</span><span class="fn-backref">(<a href="#id1">1</a>,<a href="#id3">2</a>)</span></dt>
<dd><p>Peter J. A. Cock, Christopher J. Fields, Naohisa Goto, Michael L. Heuer,
and Peter M. Rice. The Sanger FASTQ file format for sequences with quality
scores, and the Solexa/Illumina FASTQ variants. Nucl. Acids Res. (2010) 38
(6): 1767-1771. first published online December 16, 2009.
doi:10.1093/nar/gkp1137
http://nar.oxfordjournals.org/content/38/6/1767</p>
</dd>
<dt class="label" id="id6"><span class="brackets"><a class="fn-backref" href="#id2">2</a></span></dt>
<dd><p><a class="reference external" href="http://en.wikipedia.org/wiki/FASTQ_format">http://en.wikipedia.org/wiki/FASTQ_format</a></p>
</dd>
<dt class="label" id="id7"><span class="brackets"><a class="fn-backref" href="#id4">3</a></span></dt>
<dd><p><a class="reference external" href="http://www.open-bio.org/">http://www.open-bio.org/</a></p>
</dd>
</dl>
</div>
</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.4.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>