<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>skbio.alignment.StripedSmithWaterman &#8212; scikit-bio 0.5.6 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/style.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="skbio.alignment.StripedSmithWaterman.__call__" href="skbio.alignment.StripedSmithWaterman.__call__.html" />
    <link rel="prev" title="skbio.alignment.TabularMSA.write" href="skbio.alignment.TabularMSA.write.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../_static/bootstrap-sphinx.js "></script>

  </head><body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.5.6</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../io.html">File I/O (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../alignment.html">Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.tree</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.workflow</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.stats</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../metadata.html">Metadata (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.metadata</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.util</span></code>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.alignment.StripedSmithWaterman</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="section" id="skbio-alignment-stripedsmithwaterman">
<h1>skbio.alignment.StripedSmithWaterman<a class="headerlink" href="#skbio-alignment-stripedsmithwaterman" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="skbio.alignment.StripedSmithWaterman">
<em class="property">class </em><code class="sig-prename descclassname">skbio.alignment.</code><code class="sig-name descname">StripedSmithWaterman</code><a class="headerlink" href="#skbio.alignment.StripedSmithWaterman" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs a striped (banded) Smith Waterman Alignment.</p>
<p>First a StripedSmithWaterman object must be instantiated with a query
sequence. The resulting object is then callable with a target sequence and
may be reused on a large collection of target sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query_sequence</strong> (<em>string</em>) – The query sequence, this may be upper or lowercase from the set of
{A, C, G, T, N} (nucleotide) or from the set of
{A, R, N, D, C, Q, E, G, H, I, L, K, M, F, P, S, T, W, Y, V, B, Z, X, *
} (protein)</p></li>
<li><p><strong>gap_open_penalty</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – The penalty applied to creating a gap in the alignment. This CANNOT
be 0.
Default is 5.</p></li>
<li><p><strong>gap_extend_penalty</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – The penalty applied to extending a gap in the alignment. This CANNOT
be 0.
Default is 2.</p></li>
<li><p><strong>score_size</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – If your estimated best alignment score is &lt; 255 this should be 0.
If your estimated best alignment score is &gt;= 255, this should be 1.
If you don’t know, this should be 2.
Default is 2.</p></li>
<li><p><strong>mask_length</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – The distance between the optimal and suboptimal alignment ending
position &gt;= mask_length. We suggest to use len(query_sequence)/2, if
you don’t have special concerns.
Detailed description of mask_length: After locating the optimal
alignment ending position, the suboptimal alignment score can be
heuristically found by checking the second largest score in the array
that contains the maximal score of each column of the SW matrix. In
order to avoid picking the scores that belong to the alignments
sharing the partial best alignment, SSW C library masks the reference
loci nearby (mask length = mask_length) the best alignment ending
position and locates the second largest score from the unmasked
elements.
Default is 15.</p></li>
<li><p><strong>mask_auto</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – This will automatically set the used mask length to be
max(int(len(<cite>query_sequence</cite>)/2), <cite>mask_length</cite>).
Default is True.</p></li>
<li><p><strong>score_only</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – This will prevent the best alignment beginning positions (BABP) and the
cigar from being returned as a result. This overrides any setting on
<cite>score_filter</cite>, <cite>distance_filter</cite>, and <cite>override_skip_babp</cite>. It has the
highest precedence.
Default is False.</p></li>
<li><p><strong>score_filter</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – If set, this will prevent the cigar and best alignment beginning
positions (BABP) from being returned if the optimal alignment score is
less than <cite>score_filter</cite> saving some time computationally. This filter
may be overridden by <cite>score_only</cite> (prevents BABP and cigar, regardless
of other arguments), <cite>distance_filter</cite> (may prevent cigar, but will
cause BABP to be calculated), and <cite>override_skip_babp</cite> (will ensure
BABP) returned.
Default is None.</p></li>
<li><p><strong>distance_filter</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – If set, this will prevent the cigar from being returned if the length
of the <cite>query_sequence</cite> or the <cite>target_sequence</cite> is less than
<cite>distance_filter</cite> saving some time computationally. The results of
this filter may be overridden by <cite>score_only</cite> (prevents BABP and cigar,
regardless of other arguments), and <cite>score_filter</cite> (may prevent cigar).
<cite>override_skip_babp</cite> has no effect with this filter applied, as BABP
must be calculated to perform the filter.
Default is None.</p></li>
<li><p><strong>override_skip_babp</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – When True, the best alignment beginning positions (BABP) will always be
returned unless <cite>score_only</cite> is set to True.
Default is False.</p></li>
<li><p><strong>protein</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – When True, the <cite>query_sequence</cite> and <cite>target_sequence</cite> will be read as
protein sequence. When False, the <cite>query_sequence</cite> and
<cite>target_sequence</cite> will be read as nucleotide sequence. If True, a
<cite>substitution_matrix</cite> must be supplied.
Default is False.</p></li>
<li><p><strong>match_score</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – When using a nucleotide sequence, the match_score is the score added
when a match occurs. This is ignored if <cite>substitution_matrix</cite> is
provided.
Default is 2.</p></li>
<li><p><strong>mismatch_score</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – When using a nucleotide sequence, the mismatch is the score subtracted
when a mismatch occurs. This should be a negative integer.
This is ignored if <cite>substitution_matrix</cite> is provided.
Default is -3.</p></li>
<li><p><strong>substitution_matrix</strong> (<em>2D dict</em><em>, </em><em>optional</em>) – Provides the score for each possible substitution of sequence
characters. This may be used for protein or nucleotide sequences. The
entire set of possible combinations for the relevant sequence type MUST
be enumerated in the dict of dicts. This will override <cite>match_score</cite>
and <cite>mismatch_score</cite>. Required when <cite>protein</cite> is True.
Default is None.</p></li>
<li><p><strong>suppress_sequences</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, the query and target sequences will not be returned for
convenience.
Default is False.</p></li>
<li><p><strong>zero_index</strong> (<a class="reference external" href="https://docs.python.org/dev/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, all inidices will start at 0. If False, all inidices will
start at 1.
Default is True.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This is a wrapper for the SSW package <a class="footnote-reference brackets" href="#id2" id="id1">1</a>.</p>
<p><cite>mask_length</cite> has to be &gt;= 15, otherwise the suboptimal alignment
information will NOT be returned.</p>
<p><cite>match_score</cite> is a positive integer and <cite>mismatch_score</cite> is a negative
integer.</p>
<p><cite>match_score</cite> and <cite>mismatch_score</cite> are only meaningful in the context of
nucleotide sequences.</p>
<p>A substitution matrix must be provided when working with protein sequences.</p>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id2"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Zhao, Mengyao, Wan-Ping Lee, Erik P. Garrison, &amp; Gabor T.
Marth. “SSW Library: An SIMD Smith-Waterman C/C++ Library for
Applications”. PLOS ONE (2013). Web. 11 July 2014.
<a class="reference external" href="http://www.plosone.org/article/info:doi/10.1371/journal.pone.0082138">http://www.plosone.org/article/info:doi/10.1371/journal.pone.0082138</a></p>
</dd>
</dl>
<p class="rubric">Built-ins</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.StripedSmithWaterman.__call__.html#skbio.alignment.StripedSmithWaterman.__call__" title="skbio.alignment.StripedSmithWaterman.__call__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__call__</span></code></a></p></td>
<td><p>Align <cite>target_sequence</cite> to <cite>query_sequence</cite></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.StripedSmithWaterman.__hash__.html#skbio.alignment.StripedSmithWaterman.__hash__" title="skbio.alignment.StripedSmithWaterman.__hash__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__hash__</span></code></a></p></td>
<td><p>Return hash(self).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.StripedSmithWaterman.__setstate__.html#skbio.alignment.StripedSmithWaterman.__setstate__" title="skbio.alignment.StripedSmithWaterman.__setstate__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__setstate__</span></code></a></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.4.4.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>