<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>FASTA/QUAL format (skbio.io.fasta) &mdash; scikit-bio 0.2.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.1.0/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.1.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.2.1 documentation" href="../index.html" />
    <link rel="up" title="File I/O (skbio.io)" href="../io.html" />
    <link rel="next" title="FASTQ format (skbio.io.fastq)" href="skbio.io.fastq.html" />
    <link rel="prev" title="Clustal format (skbio.io.clustal)" href="skbio.io.clustal.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.1</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Sequence collections and alignments (<tt class="docutils literal"><span class="pre">skbio.alignment</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw.html">Visualizations (<tt class="docutils literal"><span class="pre">skbio.draw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../format.sequences.html">Format biological sequences (<tt class="docutils literal"><span class="pre">skbio.format.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<tt class="docutils literal"><span class="pre">skbio.diversity</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<tt class="docutils literal"><span class="pre">skbio.stats</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../parse.sequences.html">Parse biological sequences (<tt class="docutils literal"><span class="pre">skbio.parse.sequences</span></tt>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../io.html">File I/O (<tt class="docutils literal"><span class="pre">skbio.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<tt class="docutils literal"><span class="pre">skbio.util</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">FASTA/QUAL format (<tt class="docutils literal"><span class="pre">skbio.io.fasta</span></tt>)</a><ul>
<li><a class="reference internal" href="#format-support">Format Support</a></li>
<li><a class="reference internal" href="#format-specification">Format Specification</a><ul>
<li><a class="reference internal" href="#fasta-format">FASTA Format</a><ul>
<li><a class="reference internal" href="#sequence-header">Sequence Header</a></li>
<li><a class="reference internal" href="#sequence-data">Sequence Data</a></li>
</ul>
</li>
<li><a class="reference internal" href="#qual-format">QUAL Format</a></li>
</ul>
</li>
<li><a class="reference internal" href="#format-parameters">Format Parameters</a><ul>
<li><a class="reference internal" href="#qual-file-parameter-readers-and-writers">QUAL File Parameter (Readers and Writers)</a></li>
<li><a class="reference internal" href="#reader-specific-parameters">Reader-specific Parameters</a><ul>
<li><a class="reference internal" href="#generator-sequencecollection-and-alignment-reader-parameters">Generator, SequenceCollection, and Alignment Reader Parameters</a></li>
<li><a class="reference internal" href="#biologicalsequence-reader-parameters">BiologicalSequence Reader Parameters</a></li>
</ul>
</li>
<li><a class="reference internal" href="#writer-specific-parameters">Writer-specific Parameters</a></li>
</ul>
</li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#reading-and-writing-fasta-files">Reading and Writing FASTA Files</a></li>
<li><a class="reference internal" href="#reading-and-writing-fasta-qual-files">Reading and Writing FASTA/QUAL Files</a></li>
</ul>
</li>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <span class="target" id="module-skbio.io.fasta"></span><div class="section" id="fasta-qual-format-skbio-io-fasta">
<h1>FASTA/QUAL format (<a class="reference internal" href="#module-skbio.io.fasta" title="skbio.io.fasta"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.io.fasta</span></tt></a>)<a class="headerlink" href="#fasta-qual-format-skbio-io-fasta" title="Permalink to this headline">¶</a></h1>
<p>The FASTA file format (<tt class="docutils literal"><span class="pre">fasta</span></tt>) stores biological (i.e., nucleotide or
protein) sequences in a simple plain text format that is both human-readable
and easy to parse. The file format was first introduced and used in the FASTA
software package <a class="reference internal" href="#r153" id="id1">[R153]</a>. Additional descriptions of the file format can be found
in <a class="reference internal" href="#r154" id="id2">[R154]</a> and <a class="reference internal" href="#r155" id="id3">[R155]</a>.</p>
<p>An example of a FASTA-formatted file containing two DNA sequences:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;seq1 db-accession-149855
CGATGTCGATCGATCGATCGATCAG
&gt;seq2 db-accession-34989
CATCGATCGATCGATGCATGCATGCATG
</pre></div>
</div>
<p>The QUAL file format is an additional format related to FASTA. A FASTA file is
sometimes accompanied by a QUAL file, particuarly when the fasta file contains
sequences generated on a high-throughput sequencing instrument. QUAL files
store a Phred quality score (nonnegative integer) for each base in a sequence
stored in FASTA format (see <a class="reference internal" href="#r156" id="id4">[R156]</a> for more details). scikit-bio supports reading
and writing FASTA (and optionally QUAL) file formats.</p>
<div class="section" id="format-support">
<h2>Format Support<a class="headerlink" href="#format-support" title="Permalink to this headline">¶</a></h2>
<p><strong>Has Sniffer: Yes</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="8%" />
<col width="84%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Reader</th>
<th class="head">Writer</th>
<th class="head">Object Class</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Yes</td>
<td>Yes</td>
<td>generator of <a class="reference internal" href="skbio.sequence.BiologicalSequence.html#skbio.sequence.BiologicalSequence" title="skbio.sequence.BiologicalSequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.BiologicalSequence</span></tt></a> objects</td>
</tr>
<tr class="row-odd"><td>Yes</td>
<td>Yes</td>
<td><a class="reference internal" href="skbio.alignment.SequenceCollection.html#skbio.alignment.SequenceCollection" title="skbio.alignment.SequenceCollection"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.alignment.SequenceCollection</span></tt></a></td>
</tr>
<tr class="row-even"><td>Yes</td>
<td>Yes</td>
<td><a class="reference internal" href="skbio.alignment.Alignment.html#skbio.alignment.Alignment" title="skbio.alignment.Alignment"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.alignment.Alignment</span></tt></a></td>
</tr>
<tr class="row-odd"><td>Yes</td>
<td>Yes</td>
<td><a class="reference internal" href="skbio.sequence.BiologicalSequence.html#skbio.sequence.BiologicalSequence" title="skbio.sequence.BiologicalSequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.BiologicalSequence</span></tt></a></td>
</tr>
<tr class="row-even"><td>Yes</td>
<td>Yes</td>
<td><a class="reference internal" href="skbio.sequence.NucleotideSequence.html#skbio.sequence.NucleotideSequence" title="skbio.sequence.NucleotideSequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.NucleotideSequence</span></tt></a></td>
</tr>
<tr class="row-odd"><td>Yes</td>
<td>Yes</td>
<td><a class="reference internal" href="skbio.sequence.DNASequence.html#skbio.sequence.DNASequence" title="skbio.sequence.DNASequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.DNASequence</span></tt></a></td>
</tr>
<tr class="row-even"><td>Yes</td>
<td>Yes</td>
<td><a class="reference internal" href="skbio.sequence.RNASequence.html#skbio.sequence.RNASequence" title="skbio.sequence.RNASequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.RNASequence</span></tt></a></td>
</tr>
<tr class="row-odd"><td>Yes</td>
<td>Yes</td>
<td><a class="reference internal" href="skbio.sequence.ProteinSequence.html#skbio.sequence.ProteinSequence" title="skbio.sequence.ProteinSequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.ProteinSequence</span></tt></a></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">All readers and writers support an optional QUAL file via the
<tt class="docutils literal"><span class="pre">qual</span></tt> parameter. If one is provided, quality scores will be read/written
in addition to FASTA sequence data.</p>
</div>
</div>
<div class="section" id="format-specification">
<h2>Format Specification<a class="headerlink" href="#format-specification" title="Permalink to this headline">¶</a></h2>
<p>The following sections define the FASTA and QUAL file formats in detail.</p>
<div class="section" id="fasta-format">
<h3>FASTA Format<a class="headerlink" href="#fasta-format" title="Permalink to this headline">¶</a></h3>
<p>A FASTA file contains one or more biological sequences. The sequences are
stored sequentially, with a <em>record</em> for each sequence (also referred to as a
<em>FASTA record</em>). Each <em>record</em> consists of a single-line <em>header</em> (sometimes
referred to as a <em>defline</em>, <em>label</em>, <em>description</em>, or <em>comment</em>) followed by
the sequence data, optionally split over multiple lines. Blank or
whitespace-only lines are not allowed anywhere in the FASTA file.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">scikit-bio does not currently support legacy FASTA format (i.e.,
headers/comments denoted with a semicolon). The format supported by
scikit-bio (described below in detail) most closely resembles the
description given in NCBI&#8217;s BLAST documentation <a class="reference internal" href="#r155" id="id5">[R155]</a>. See <a class="reference internal" href="#r154" id="id6">[R154]</a> for more
details on legacy FASTA format. If you would like legacy FASTA format
support added to scikit-bio, please consider submitting a feature request on
the
<a class="reference external" href="https://github.com/biocore/scikit-bio/issues">scikit-bio issue tracker</a>
(pull requests are also welcome!).</p>
</div>
<div class="section" id="sequence-header">
<h4>Sequence Header<a class="headerlink" href="#sequence-header" title="Permalink to this headline">¶</a></h4>
<p>Each sequence header consists of a single line beginning with a greater-than
(<tt class="docutils literal"><span class="pre">&gt;</span></tt>) symbol. Immediately following this is a sequence identifier (ID) and
description separated by one or more whitespace characters. Both sequence ID
and description are optional and are represented as the empty string (<tt class="docutils literal"><span class="pre">''</span></tt>)
in scikit-bio&#8217;s objects if they are not present in the header.</p>
<p>A sequence ID consists of a single <em>word</em>: all characters after the greater-
than symbol and before the first whitespace character (if any) are taken as the
sequence ID. Unique sequence IDs are not strictly enforced by the FASTA format
itself. A single standardized ID format is similarly not enforced by FASTA
format, though it is often common to use a unique library accession number for
a sequence ID (e.g., NCBI&#8217;s FASTA defline format <a class="reference internal" href="#r157" id="id7">[R157]</a>).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">scikit-bio will enforce sequence ID uniqueness depending on the type
of object that the FASTA file is read into. For example, reading a FASTA
file as a generator of <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> objects will not enforce
unique IDs since it simply yields each sequence it finds in the FASTA file.
However, if the FASTA file is read into a <tt class="docutils literal"><span class="pre">SequenceCollection</span></tt> object, ID
uniqueness will be enforced because that is a requirement of a
<tt class="docutils literal"><span class="pre">SequenceCollection</span></tt>.</p>
</div>
<p>If a description is present, it is taken as the remaining characters that
follow the sequence ID and initial whitespace(s). The description is considered
additional information about the sequence (e.g., comments about the source of
the sequence or the molecule that it encodes).</p>
<p>For example, consider the following header:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;seq1 db-accession-149855
</pre></div>
</div>
<p><tt class="docutils literal"><span class="pre">seq1</span></tt> is the sequence ID and <tt class="docutils literal"><span class="pre">db-accession-149855</span></tt> is the sequence
description.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">scikit-bio&#8217;s readers will remove all leading and trailing whitespace
from the description. If a header line begins with whitespace following the
<tt class="docutils literal"><span class="pre">&gt;</span></tt>, the ID is assumed to be missing and the remainder of the line is
taken as the description.</p>
</div>
</div>
<div class="section" id="sequence-data">
<h4>Sequence Data<a class="headerlink" href="#sequence-data" title="Permalink to this headline">¶</a></h4>
<p>Biological sequence data follows the header, and can be split over multiple
lines. The sequence data (i.e., nucleotides or amino acids) are stored using
the standard IUPAC lexicon (single-letter codes).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>scikit-bio supports both upper and lower case characters. Both <tt class="docutils literal"><span class="pre">-</span></tt>
and <tt class="docutils literal"><span class="pre">.</span></tt> are supported as gap characters. See <a class="reference internal" href="../sequence.html#module-skbio.sequence" title="skbio.sequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence</span></tt></a> for
more details on how scikit-bio interprets sequence data in its in-memory
objects.</p>
<p class="last">scikit-bio will remove leading and trailing whitespace from each line of
sequence data before joining the sequence chunks into a single sequence.
Whitespace characters are <strong>not</strong> removed from the middle of the sequence
chunks. Likewise, other invalid IUPAC characters are <strong>not</strong> removed from
the sequence data as it is read. Thus, it is possible to create an invalid
in-memory sequence object (see warning below).</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">In an effort to maintain reasonable performance while reading
FASTA files (which can be quite large), validation of sequence data is
<strong>not</strong> performed during reading. It is the responsibility of the user to
validate their in-memory representation of the data if desired (e.g., by
calling <tt class="docutils literal"><span class="pre">is_valid</span></tt> on the returned object). Thus, it is possible to read
invalid characters into objects (e.g. whitespace occurring in the middle of
a sequence, or invalid IUPAC DNA characters in a DNA sequence).</p>
</div>
</div>
</div>
<div class="section" id="qual-format">
<h3>QUAL Format<a class="headerlink" href="#qual-format" title="Permalink to this headline">¶</a></h3>
<p>A QUAL file contains quality scores for one or more biological sequences stored
in a corresponding FASTA file. QUAL format is very similar to FASTA format: it
stores records sequentially, with each record beginning with a header line
containing a sequence ID and description. The same rules apply to QUAL headers
as FASTA headers (see the above sections for details). scikit-bio processes
FASTA and QUAL headers in exactly the same way.</p>
<p>Instead of storing biological sequence data in each record, a QUAL file stores
a Phred quality score for each base in the corresponding sequence. Quality
scores are represented as nonnegative integers separated by whitespace
(typically a single space or newline), and can span multiple lines.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>When reading FASTA and QUAL files, scikit-bio requires records to be
in the same order in both files (i.e., each FASTA and QUAL record must have
the same ID and description after being parsed). In addition to having the
same order, the number of FASTA records must match the number of QUAL
records (i.e., missing or additonal records are not allowed). scikit-bio
also requires that the number of quality scores match the number of bases in
the corresponding sequence.</p>
<p class="last">When writing FASTA and QUAL files, scikit-bio will maintain the same
ordering of records in both files (i.e., using the same ID and description
in both records) to support future reading.</p>
</div>
</div>
</div>
<div class="section" id="format-parameters">
<h2>Format Parameters<a class="headerlink" href="#format-parameters" title="Permalink to this headline">¶</a></h2>
<p>The following parameters are available to change how FASTA/QUAL files are read
or written in scikit-bio.</p>
<div class="section" id="qual-file-parameter-readers-and-writers">
<h3>QUAL File Parameter (Readers and Writers)<a class="headerlink" href="#qual-file-parameter-readers-and-writers" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">qual</span></tt> parameter is available to all FASTA format readers and writers. It
can be any file-like type supported by scikit-bio&#8217;s I/O registry (e.g., file
handle, file path, etc.). If <tt class="docutils literal"><span class="pre">qual</span></tt> is provided when reading, quality scores
will be included in each in-memory <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> object, in addition
to sequence data stored in the FASTA file. When writing, quality scores will be
written in QUAL format in addition to the sequence data being written in FASTA
format.</p>
</div>
<div class="section" id="reader-specific-parameters">
<h3>Reader-specific Parameters<a class="headerlink" href="#reader-specific-parameters" title="Permalink to this headline">¶</a></h3>
<p>The available reader parameters differ depending on which reader is used.</p>
<div class="section" id="generator-sequencecollection-and-alignment-reader-parameters">
<h4>Generator, SequenceCollection, and Alignment Reader Parameters<a class="headerlink" href="#generator-sequencecollection-and-alignment-reader-parameters" title="Permalink to this headline">¶</a></h4>
<p>The <tt class="docutils literal"><span class="pre">constructor</span></tt> parameter can be used with the <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt>
generator, <tt class="docutils literal"><span class="pre">SequenceCollection</span></tt>, and <tt class="docutils literal"><span class="pre">Alignment</span></tt> FASTA readers.
<tt class="docutils literal"><span class="pre">constructor</span></tt> specifies the in-memory type of each sequence that is parsed,
and defaults to <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt>. <tt class="docutils literal"><span class="pre">constructor</span></tt> should be a subclass of
<tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt>. For example, if you know that the FASTA file you&#8217;re
reading contains protein sequences, you would pass
<tt class="docutils literal"><span class="pre">constructor=ProteinSequence</span></tt> to the reader call.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The FASTA sniffer will not attempt to guess the <tt class="docutils literal"><span class="pre">constructor</span></tt>
parameter, so it will always default to <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> if another
type is not provided to the reader.</p>
</div>
</div>
<div class="section" id="biologicalsequence-reader-parameters">
<h4>BiologicalSequence Reader Parameters<a class="headerlink" href="#biologicalsequence-reader-parameters" title="Permalink to this headline">¶</a></h4>
<p>The <tt class="docutils literal"><span class="pre">seq_num</span></tt> parameter can be used with the <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt>,
<tt class="docutils literal"><span class="pre">NucleotideSequence</span></tt>, <tt class="docutils literal"><span class="pre">DNASequence</span></tt>, <tt class="docutils literal"><span class="pre">RNASequence</span></tt>, and
<tt class="docutils literal"><span class="pre">ProteinSequence</span></tt> FASTA readers. <tt class="docutils literal"><span class="pre">seq_num</span></tt> specifies which sequence to read
from the FASTA file (and optional QUAL file), and defaults to 1 (i.e., such
that the first sequence is read). For example, to read the 50th sequence from a
FASTA file, you would pass <tt class="docutils literal"><span class="pre">seq_num=50</span></tt> to the reader call.</p>
</div>
</div>
<div class="section" id="writer-specific-parameters">
<h3>Writer-specific Parameters<a class="headerlink" href="#writer-specific-parameters" title="Permalink to this headline">¶</a></h3>
<p>The following parameters are available to all FASTA format writers:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">id_whitespace_replacement</span></tt>: string to replace <strong>each</strong> whitespace
character in a sequence ID. This parameter is useful for cases where an
in-memory sequence ID contains whitespace, which would result in an on-disk
representation that would not be read back into memory as the same ID (since
IDs in FASTA format cannot contain whitespace). Defaults to <tt class="docutils literal"><span class="pre">_</span></tt>. If
<tt class="docutils literal"><span class="pre">None</span></tt>, no whitespace replacement is performed and IDs are written as they
are stored in memory (this has the potential to create an invalid
FASTA-formatted file; see note below). This parameter also applies to a QUAL
file if one is provided.</li>
<li><tt class="docutils literal"><span class="pre">description_newline_replacement</span></tt>: string to replace <strong>each</strong> newline
character in a sequence description. Since a FASTA header must be a single
line, newlines are not allowed in sequence descriptions and must be replaced
in order to write a valid FASTA file. Defaults to a single space. If
<tt class="docutils literal"><span class="pre">None</span></tt>, no newline replacement is performed and descriptions are written as
they are stored in memory (this has the potential to create an invalid
FASTA-formatted file; see note below). This parameter also applies to a QUAL
file if one is provided.</li>
<li><tt class="docutils literal"><span class="pre">max_width</span></tt>: integer specifying the maximum line width (i.e., number of
characters) for sequence data and/or quality scores. If a sequence or its
quality scores are longer than <tt class="docutils literal"><span class="pre">max_width</span></tt>, it will be split across
multiple lines, each with a maximum width of <tt class="docutils literal"><span class="pre">max_width</span></tt>. Note that there
are some caveats when splitting quality scores. A single quality score will
<em>never</em> be split across multiple lines, otherwise it would become two
different quality scores when read again. Thus, splitting only occurs
<em>between</em> quality scores. This makes it possible to have a single long
quality score written on its own line that exceeds <tt class="docutils literal"><span class="pre">max_width</span></tt>. For
example, the quality score <tt class="docutils literal"><span class="pre">12345</span></tt> would not be split across multiple lines
even if <tt class="docutils literal"><span class="pre">max_width=3</span></tt>. Thus, a 5-character line would be written. Default
behavior is to not split sequence data or quality scores across multiple
lines.</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The FASTA format writers will have noticeably better runtime
performance if <tt class="docutils literal"><span class="pre">id_whitespace_replacement</span></tt> and/or
<tt class="docutils literal"><span class="pre">description_newline_replacement</span></tt> are set to <tt class="docutils literal"><span class="pre">None</span></tt> so that whitespace
replacement is not performed during writing. However, this can potentially
create invalid FASTA files, especially if there are newline characters in
the IDs or descriptions. For IDs with whitespace, this can also affect how
the IDs are read into memory in a subsequent read operation. For example, if
an in-memory sequence ID is <tt class="docutils literal"><span class="pre">'seq</span> <span class="pre">1'</span></tt> and
<tt class="docutils literal"><span class="pre">id_whitespace_replacement=None</span></tt>, reading the FASTA file back into memory
would result in an ID of <tt class="docutils literal"><span class="pre">'seq'</span></tt>, and <tt class="docutils literal"><span class="pre">'1'</span></tt> would be part of the
sequence description.</p>
</div>
</div>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<div class="section" id="reading-and-writing-fasta-files">
<h3>Reading and Writing FASTA Files<a class="headerlink" href="#reading-and-writing-fasta-files" title="Permalink to this headline">¶</a></h3>
<p>Suppose we have the following FASTA file with five equal-length sequences
(example modified from <a class="reference internal" href="#r158" id="id8">[R158]</a>):</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;seq1 Turkey
AAGCTNGGGCATTTCAGGGTGAGCCCGGGCAATACAGGGTAT
&gt;seq2 Salmo gair
AAGCCTTGGCAGTGCAGGGTGAGCCGTGG
CCGGGCACGGTAT
&gt;seq3 H. Sapiens
ACCGGTTGGCCGTTCAGGGTACAGGTTGGCCGTTCAGGGTAA
&gt;seq4 Chimp
AAACCCTTGCCG
TTACGCTTAAAC
CGAGGCCGGGAC
ACTCAT
&gt;seq5 Gorilla
AAACCCTTGCCGGTACGCTTAAACCATTGCCGGTACGCTTAA
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Original copyright notice for the above example file:</p>
<p class="last"><em>(c) Copyright 1986-2008 by The University of Washington. Written by Joseph
Felsenstein. Permission is granted to copy this document provided that no
fee is charged for it and that this copyright notice is not removed.</em></p>
</div>
<p>Note that the sequences are not required to be of equal length in order for the
file to be a valid FASTA file (this depends on the object that you&#8217;re reading
the file into). Also note that some of the sequences occur on a single line,
while others are split across multiple lines.</p>
<p>Let&#8217;s define this file in-memory as a <tt class="docutils literal"><span class="pre">StringIO</span></tt>, though this could be a real
file path, file handle, or anything that&#8217;s supported by scikit-bio&#8217;s I/O
registry in practice:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">StringIO</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fs</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="s">&quot;&gt;seq1 Turkey</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;AAGCTNGGGCATTTCAGGGTGAGCCCGGGCAATACAGGGTAT</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;&gt;seq2 Salmo gair</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;AAGCCTTGGCAGTGCAGGGTGAGCCGTGG</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;CCGGGCACGGTAT</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;&gt;seq3 H. Sapiens</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;ACCGGTTGGCCGTTCAGGGTACAGGTTGGCCGTTCAGGGTAA</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;&gt;seq4 Chimp</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;AAACCCTTGCCG</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;TTACGCTTAAAC</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;CGAGGCCGGGAC</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;ACTCAT</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;&gt;seq5 Gorilla</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;AAACCCTTGCCGGTACGCTTAAACCATTGCCGGTACGCTTAA</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span>
</pre></div>
</div>
<p>Let&#8217;s read the FASTA file into a <tt class="docutils literal"><span class="pre">SequenceCollection</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">SequenceCollection</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span> <span class="o">=</span> <span class="n">SequenceCollection</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">sequence_lengths</span><span class="p">()</span>
<span class="go">[42, 42, 42, 42, 42]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
<span class="go">[&#39;seq1&#39;, &#39;seq2&#39;, &#39;seq3&#39;, &#39;seq4&#39;, &#39;seq5&#39;]</span>
</pre></div>
</div>
<p>We see that all 5 sequences have 42 characters, and that each of the sequence
IDs were successfully read into memory.</p>
<p>Since these sequences are of equal length (presumably because they&#8217;ve been
aligned), let&#8217;s load the FASTA file into an <tt class="docutils literal"><span class="pre">Alignment</span></tt> object, which is a
more appropriate data structure:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">Alignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span> <span class="c"># reload the StringIO to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">()</span>
<span class="go">42</span>
</pre></div>
</div>
<p>Note that we were able to read the FASTA file into two different data
structures (<tt class="docutils literal"><span class="pre">SequenceCollection</span></tt> and <tt class="docutils literal"><span class="pre">Alignment</span></tt>) using the exact same
<tt class="docutils literal"><span class="pre">read</span></tt> method call (and underlying reading/parsing logic). Also note that we
didn&#8217;t specify a file format in the <tt class="docutils literal"><span class="pre">read</span></tt> call. The FASTA sniffer detected
the correct file format for us!</p>
<p>Let&#8217;s inspect the type of sequences stored in the <tt class="docutils literal"><span class="pre">Alignment</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&lt;BiologicalSequence: AAGCTNGGGC... (length: 42)&gt;</span>
</pre></div>
</div>
<p>By default, sequences are loaded as <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> objects. We can
change the type of sequence via the <tt class="docutils literal"><span class="pre">constructor</span></tt> parameter:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">DNASequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span> <span class="c"># reload the StringIO to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">constructor</span><span class="o">=</span><span class="n">DNASequence</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&lt;DNASequence: AAGCTNGGGC... (length: 42)&gt;</span>
</pre></div>
</div>
<p>We now have an <tt class="docutils literal"><span class="pre">Alignment</span></tt> of <tt class="docutils literal"><span class="pre">DNASequence</span></tt> objects instead of
<tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> objects. Validation of sequence character data is not
performed during reading (see warning above for details). To verify that each
of the sequences are valid DNA sequences:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span><span class="o">.</span><span class="n">is_valid</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
<p>To write the alignment in FASTA format:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">new_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_fh</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">new_fh</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">&gt;seq1 Turkey</span>
<span class="go">AAGCTNGGGCATTTCAGGGTGAGCCCGGGCAATACAGGGTAT</span>
<span class="go">&gt;seq2 Salmo gair</span>
<span class="go">AAGCCTTGGCAGTGCAGGGTGAGCCGTGGCCGGGCACGGTAT</span>
<span class="go">&gt;seq3 H. Sapiens</span>
<span class="go">ACCGGTTGGCCGTTCAGGGTACAGGTTGGCCGTTCAGGGTAA</span>
<span class="go">&gt;seq4 Chimp</span>
<span class="go">AAACCCTTGCCGTTACGCTTAAACCGAGGCCGGGACACTCAT</span>
<span class="go">&gt;seq5 Gorilla</span>
<span class="go">AAACCCTTGCCGGTACGCTTAAACCATTGCCGGTACGCTTAA</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">new_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Both <tt class="docutils literal"><span class="pre">SequenceCollection</span></tt> and <tt class="docutils literal"><span class="pre">Alignment</span></tt> load all of the sequences from
the FASTA file into memory at once. If the FASTA file is large (which is often
the case), this may be infeasible if you don&#8217;t have enough memory. To work
around this issue, you can stream the sequences using scikit-bio&#8217;s
generator-based FASTA reader and writer. The generator-based reader yields
<tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> objects (or subclasses if <tt class="docutils literal"><span class="pre">constructor</span></tt> is supplied)
one at a time, instead of loading all sequences into memory. For example, let&#8217;s
use the generator-based reader to process a single sequence at a time in a
<tt class="docutils literal"><span class="pre">for</span></tt> loop:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">skbio.io</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span> <span class="c"># reload the StringIO to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">skbio</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">format</span><span class="o">=</span><span class="s">&#39;fasta&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">seq</span>
<span class="go">&lt;BiologicalSequence: AAGCTNGGGC... (length: 42)&gt;</span>
<span class="go">&lt;BiologicalSequence: AAGCCTTGGC... (length: 42)&gt;</span>
<span class="go">&lt;BiologicalSequence: ACCGGTTGGC... (length: 42)&gt;</span>
<span class="go">&lt;BiologicalSequence: AAACCCTTGC... (length: 42)&gt;</span>
<span class="go">&lt;BiologicalSequence: AAACCCTTGC... (length: 42)&gt;</span>
</pre></div>
</div>
<p>A single sequence can also be read into a <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> (or subclass):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">BiologicalSequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span> <span class="c"># reload the StringIO to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">BiologicalSequence</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>
<span class="go">&lt;BiologicalSequence: AAGCTNGGGC... (length: 42)&gt;</span>
</pre></div>
</div>
<p>By default, the first sequence in the FASTA file is read. This can be
controlled with <tt class="docutils literal"><span class="pre">seq_num</span></tt>. For example, to read the fifth sequence:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span> <span class="c"># reload the StringIO to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">BiologicalSequence</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">seq_num</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="go">&lt;BiologicalSequence: AAACCCTTGC... (length: 42)&gt;</span>
</pre></div>
</div>
<p>We can use the same API to read the fifth sequence into a <tt class="docutils literal"><span class="pre">DNASequence</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span> <span class="c"># reload the StringIO to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna_seq</span> <span class="o">=</span> <span class="n">DNASequence</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">seq_num</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna_seq</span>
<span class="go">&lt;DNASequence: AAACCCTTGC... (length: 42)&gt;</span>
</pre></div>
</div>
<p>Individual sequence objects can also be written in FASTA format:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">new_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna_seq</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_fh</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">new_fh</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">&gt;seq5 Gorilla</span>
<span class="go">AAACCCTTGCCGGTACGCTTAAACCATTGCCGGTACGCTTAA</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">new_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="reading-and-writing-fasta-qual-files">
<h3>Reading and Writing FASTA/QUAL Files<a class="headerlink" href="#reading-and-writing-fasta-qual-files" title="Permalink to this headline">¶</a></h3>
<p>In addition to reading and writing standalone FASTA files, scikit-bio also
supports reading and writing FASTA and QUAL files together. Suppose we have the
following FASTA file:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;seq1 db-accession-149855
CGATGTC
&gt;seq2 db-accession-34989
CATCG
</pre></div>
</div>
<p>Also suppose we have the following QUAL file:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;seq1 db-accession-149855
40 39 39 4
50 1 100
&gt;seq2 db-accession-34989
3 3 10 42 80
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fasta_fs</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="s">&quot;&gt;seq1 db-accession-149855</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;CGATGTC</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;&gt;seq2 db-accession-34989</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;CATCG</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fasta_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fasta_fs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">qual_fs</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="s">&quot;&gt;seq1 db-accession-149855</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;40 39 39 4</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;50 1 100</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;&gt;seq2 db-accession-34989</span><span class="se">\n</span><span class="s">&quot;</span>
<span class="gp">... </span>    <span class="s">&quot;3 3 10 42 80</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">qual_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">qual_fs</span><span class="p">)</span>
</pre></div>
</div>
<p>To read in a single <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> at a time, we can use the
generator-based reader as we did above, providing both FASTA and QUAL files:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">skbio</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fasta_fh</span><span class="p">,</span> <span class="n">qual</span><span class="o">=</span><span class="n">qual_fh</span><span class="p">,</span> <span class="n">format</span><span class="o">=</span><span class="s">&#39;fasta&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">seq</span>
<span class="gp">... </span>    <span class="n">seq</span><span class="o">.</span><span class="n">quality</span>
<span class="go">&lt;BiologicalSequence: CGATGTC (length: 7)&gt;</span>
<span class="go">array([ 40,  39,  39,   4,  50,   1, 100])</span>
<span class="go">&lt;BiologicalSequence: CATCG (length: 5)&gt;</span>
<span class="go">array([ 3,  3, 10, 42, 80])</span>
</pre></div>
</div>
<p>Note that the sequence objects have quality scores since we provided a QUAL
file. The other FASTA readers operate in a similar manner.</p>
<p>Now let&#8217;s load the sequences and their quality scores into a
<tt class="docutils literal"><span class="pre">SequenceCollection</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fasta_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fasta_fs</span><span class="p">)</span> <span class="c"># reload to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">qual_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">qual_fs</span><span class="p">)</span> <span class="c"># reload to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span> <span class="o">=</span> <span class="n">SequenceCollection</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fasta_fh</span><span class="p">,</span> <span class="n">qual</span><span class="o">=</span><span class="n">qual_fh</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span>
<span class="go">&lt;SequenceCollection: n=2; mean +/- std length=6.00 +/- 1.00&gt;</span>
</pre></div>
</div>
<p>To write the sequence data and quality scores in the <tt class="docutils literal"><span class="pre">SequenceCollection</span></tt> to
FASTA and QUAL files, respectively, we run:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">new_fasta_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_qual_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_fasta_fh</span><span class="p">,</span> <span class="n">qual</span><span class="o">=</span><span class="n">new_qual_fh</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">new_fasta_fh</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">&gt;seq1 db-accession-149855</span>
<span class="go">CGATGTC</span>
<span class="go">&gt;seq2 db-accession-34989</span>
<span class="go">CATCG</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">new_qual_fh</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">&gt;seq1 db-accession-149855</span>
<span class="go">40 39 39 4 50 1 100</span>
<span class="go">&gt;seq2 db-accession-34989</span>
<span class="go">3 3 10 42 80</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">new_fasta_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_qual_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<table class="docutils citation" frame="void" id="r153" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[R153]</a></td><td>Lipman, DJ; Pearson, WR (1985). &#8220;Rapid and sensitive protein similarity
searches&#8221;. Science 227 (4693): 1435-41.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r154" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R154]</td><td><em>(<a class="fn-backref" href="#id2">1</a>, <a class="fn-backref" href="#id6">2</a>)</em> <a class="reference external" href="http://en.wikipedia.org/wiki/FASTA_format">http://en.wikipedia.org/wiki/FASTA_format</a></td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r155" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R155]</td><td><em>(<a class="fn-backref" href="#id3">1</a>, <a class="fn-backref" href="#id5">2</a>)</em> <a class="reference external" href="http://blast.ncbi.nlm.nih.gov/blastcgihelp.shtml">http://blast.ncbi.nlm.nih.gov/blastcgihelp.shtml</a></td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r156" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id4">[R156]</a></td><td><a class="reference external" href="https://www.broadinstitute.org/crd/wiki/index.php/Qual">https://www.broadinstitute.org/crd/wiki/index.php/Qual</a></td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r157" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id7">[R157]</a></td><td>Madden T. The BLAST Sequence Analysis Tool. 2002 Oct 9
[Updated 2003 Aug 13]. In: McEntyre J, Ostell J, editors. The NCBI Handbook
[Internet]. Bethesda (MD): National Center for Biotechnology Information
(US); 2002-. Chapter 16. Available from:
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK21097/">http://www.ncbi.nlm.nih.gov/books/NBK21097/</a></td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="r158" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id8">[R158]</a></td><td><a class="reference external" href="http://evolution.genetics.washington.edu/phylip/doc/sequence.html">http://evolution.genetics.washington.edu/phylip/doc/sequence.html</a></td></tr>
</tbody>
</table>
</div>
</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>